---
title: "Determine best model fits"
author: "Alex Cope"
date: '2023-05-02'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ape)
library(cowplot)
library(ggtree)
library(scales)
suppressMessages(source("helperFunctions.R"))

cleanSpeciesLabels <- function(x)
{
   y <- x %>% str_remove(".max.cds") %>% 
          str_extract("[a-z]+_[a-z]+") %>%
          str_split("_") %>% 
          lapply(function(x){paste0(toupper(substring(x[1],1,1)),". ",x[2])})
          #lapply(function(x){paste(str_to_sentence(x[1]),x[2])})
  unlist(y)
}


cleanTree <- function(tree,target.species)
{
  tips.to.drop <- tree$tip.label[which(!(tree$tip.label %in% target.species))]
  cleaned <- drop.tip(tree,tips.to.drop)
  return(cleaned)
}


cleanColumnNames <- function(col_name,list_of_option)
{
  ## find potential column name from list of options
  index <- unlist(purrr::map(col_name, 
            function(x){
              ## this species will match "metschnikowia_matae_maris.max.cds" and "metschnikowia_matae.max.cds", so force it to match the latter
              if (x == "metschnikowia_matae")
                {
                  x <- paste0(x,".max.cds")
                }
              str_which(list_of_option,x)
            }
    ))
  return(list_of_option[index])
}

createEmpvsROCdf <- function(target.species,k1.phi,k2.phi,dist.matrix,orthogroups,empirical.data)
{
  dist.from.target <- dist.matrix[,target.species]
  closest.species <- names(dist.from.target[which.min(dist.from.target)])
  
  k1.phi.target <- k1.phi %>% 
    dplyr::select(Orthogroups,!!as.name(target.species)) %>%
    dplyr::rename(Phi_K1 = !!as.name(target.species))
  k2.phi.target <- k2.phi %>% 
    dplyr::select(Orthogroups,!!as.name(target.species)) %>%
    dplyr::rename(Phi_K2 = !!as.name(target.species))
  
  
  ortho.species <- orthogroups %>% 
    dplyr::select(Orthogroups,!!as.name(target.species),!!as.name(closest.species))
  ortho.species <- ortho.species %>% 
    left_join(empirical.data[[closest.species]],
              by=setNames("GeneID",closest.species )
              )
  ortho.species <- ortho.species %>% 
    left_join(k1.phi.target,by="Orthogroups") %>%
    left_join(k2.phi.target,by="Orthogroups") %>%
    dplyr::select(Orthogroups,Empirical,Phi_K1,Phi_K2) %>%
    filter(!is.na(Empirical) & !is.na(Phi_K1)) 
  return(ortho.species)
}

comparePhiEmpAcrossSpecies <- function(target.species,k1.phi,k2.phi,dist.matrix,orthogroups,empirical.data)
{
  dist.from.target <- dist.matrix[,target.species]
  closest.species <- names(dist.from.target[which.min(dist.from.target)])
  
  ortho.species <- createEmpvsROCdf(target.species,k1.phi,k2.phi,dist.matrix,orthogroups,empirical.data)
  
  ortho.species.long <- ortho.species %>% 
    pivot_longer(starts_with("Phi"),names_to = "Model",values_to="Phi")
  
  across.species.corr <- ortho.species.long %>% 
    group_by(Model) %>% 
    summarize(cor.test(Empirical,Phi,method="spearman",use="complete.obs") %>% tidy()) %>% 
    dplyr::select(Model,estimate,p.value) %>%
    pivot_wider(values_from = c(estimate,p.value),names_from = Model) %>%
    mutate(Compared.to = closest.species,
           Divergence = unname(dist.from.target[which.min(dist.from.target)]) )
  return(across.species.corr)
}


plotSpecies <- function(species.df,cub=c("Phi_Shared_Mutation","Phi_Variable_Mutation"))
{
  title <- cleanSpeciesLabels(unique(species.df$Species))
  species.df <- species.df %>% filter(CUB %in% cub)
  species.df.wide <- species.df %>% pivot_wider(id_cols = c("Species","GeneID","Empirical"),names_from="CUB",values_from = "Predicted")
  species.df$CUB <- factor(species.df$CUB,levels=c(cub[1],cub[2]))
  levels(species.df$CUB) <- c("ConstMut","VarMut")
  emp.vs.pred <- ggplot(species.df,aes(x=Empirical,y=Predicted)) + 
               geom_point(size=0.5,show.legend = F) + 
               scale_x_log10(labels = trans_format("log10", math_format(10^.x))) + 
               scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
               theme_cowplot() +
               stat_cor(method="spearman",label.sep = "\n",label.x.npc = "left",label.y.npc = "top",show.legend = F) +
               stat_cor(mapping=aes(x=Empirical,y=Predicted),method="spearman",label.sep = "\n",label.x.npc = "left",label.y.npc = "bottom",inherit.aes=F) +
               facet_wrap(~CUB) +
               ggtitle(title) +
               theme(aspect.ratio=1)
  k1.vs.k2 <- ggplot(species.df.wide,aes(x=!!as.name(cub[1]),y=!!as.name(cub[2]))) + 
              geom_point(size=0.5) +
              scale_x_log10(labels = trans_format("log10", math_format(10^.x))) + 
              scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
              theme_cowplot() +
              #scale_color_manual(values = c("#E41A1C","#377EB8"),labels = c("Lower GC3% ","Higher GC3%"),name = "Gene Cluster") + 
              xlab("Predicted (ConstMut)") +
              ylab("Predicted (VarMut)") +
              stat_cor(aes(x=!!as.name(cub[1]),y=!!as.name(cub[2])),method="spearman",label.sep = "\n",label.x.npc = "left",label.y.npc = "top",show.legend = F,inherit.aes = F) +
              theme(legend.position="top",legend.text = element_text(size=10),legend.title = element_text(size=12)) +
               theme(aspect.ratio=1)
  
  plot_grid(emp.vs.pred,k1.vs.k2,ncol=2,align = "h", axis = "bt",rel_widths = c(1,0.5))
}


```

# Set up color schemes

```{r}
scale_colour_brewer_d <- function(..., palette = "Set3") {
  scale_colour_brewer(..., palette = palette )
}

scale_fill_brewer_d <- function(..., palette = "Set1") {
  scale_fill_brewer(..., palette = palette)
  
}

scale_colour_brewer_c <- function(..., palette = "BuGn") {
  scale_colour_distiller(..., palette = palette )
}

scale_fill_brewer_c <- function(..., palette = "BuGn") {
  scale_fill_distiller(..., palette = palette)
}

options(
  ggplot2.discrete.colour = scale_colour_brewer_d,
  ggplot2.discrete.fill = scale_fill_brewer_d,
  ggplot2.continuous.colour = scale_colour_brewer_c,
  ggplot2.continuous.fill = scale_fill_brewer_c
)
```

```{r warning = F}

fungi.tree <- read.tree("../Data/tree_with_cds_labels.nwk")
all.species <- readLines("../Data/all_fungi.txt")
fungi.tree <- cleanTree(fungi.tree,all.species)
dist.matrix <- cophenetic.phylo(fungi.tree)

species.w.emp <- list.files("../Data/Expression/",full.names = F,pattern=".max.cds")
empirical.data.file <- list.files("../Data/Expression/",full.names = T,include.dirs = F,pattern=".max.cds")
names(empirical.data.file) <- species.w.emp


dist.matrix <- dist.matrix[species.w.emp,]

empirical.data <- purrr::map(empirical.data.file,function(x) {
  read_csv(x,col_types = cols()) %>% 
    mutate(across(contains("tpm_kallisto"),~na_if(., 0))) %>%
    mutate(across(contains("tpm_kallisto"),log)) %>%
    mutate(
      Log.Mean.Expression=rowMeans(dplyr::select(.,contains("tpm_kallisto")),na.rm=T),
      Empirical=exp(Log.Mean.Expression),
      GeneID=str_remove(GeneID,"-mRNA-1")) %>% 
      dplyr::select(-contains("tpm_kallisto"),-Log.Mean.Expression)
    })

k1.phi <- read_csv("../Post_analysis/phi_k1_shen_etal_ortholog_matrix_2022_07_08.tsv",col_types = cols())  %>% 
       column_to_rownames("Orthogroup") %>%
       rename_with(~cleanColumnNames(.x,fungi.tree$tip.label)) %>%
       rownames_to_column("Orthogroups")

k2.phi <- read_csv("../Post_analysis/phi_k2_selectionShared_shen_etal_ortholog_matrix_2022_07_08.tsv",col_types = cols())  %>% 
       column_to_rownames("Orthogroup") %>%
       rename_with(~cleanColumnNames(.x,fungi.tree$tip.label)) %>%
       rownames_to_column("Orthogroups")


orthogroups <- read_tsv("../Data/shen_etal_ortholog_matrix_2021_12_17.tsv",col_types = cols())
orthogroups <- orthogroups %>% 
       column_to_rownames("Orthogroup") %>%
       rename_with(~cleanColumnNames(.x,fungi.tree$tip.label)) %>%
       rownames_to_column("Orthogroups")

names(all.species) <- all.species


roc.emp.df <- lapply(all.species,createEmpvsROCdf,
                      k1.phi = k1.phi,
                      k2.phi = k2.phi,
                      dist.matrix = dist.matrix,
                      orthogroups = orthogroups,
                      empirical.data = empirical.data) %>% 
  bind_rows(.id="Species")


compare.emp <- lapply(all.species,comparePhiEmpAcrossSpecies,
                      k1.phi = k1.phi,
                      k2.phi = k2.phi,
                      dist.matrix = dist.matrix,
                      orthogroups = orthogroups,
                      empirical.data = empirical.data) %>% 
  bind_rows(.id="Species")

clades <- read_tsv("../Data/clades.txt")
clades <- clades %>% 
  dplyr::slice(match(fungi.tree$tip.label,file_name_id)) %>%
  dplyr::rename(Species=file_name_id,Clade=Major.clade)


compare.emp <- compare.emp %>% 
  left_join(clades,by="Species")


```



```{r}

compare.emp <- compare.emp %>%
  mutate(Diff = estimate_Phi_K2 - estimate_Phi_K1,
    Improvement = Diff/abs(estimate_Phi_K1)) 
compare.emp.improved <- compare.emp %>%
  filter(Improvement >= 0.25 & Diff > 0 & estimate_Phi_K2 > 0)
improved.fits.species <- compare.emp.improved %>%
  dplyr::select(Species) %>%
  deframe()

write(improved.fits.species,"../Data/improved_species_by_gene_expression_25_percent_2023_05_03.txt",ncolumns = 1)

compare.emp.unimproved <- compare.emp %>%
  filter(Improvement < 0.25 | (Improvement >= 0.25 & estimate_Phi_K2 < 0))



compare.emp <- compare.emp %>% 
  mutate(estimate_Phi = ifelse(Species %in% improved.fits.species,estimate_Phi_K2,estimate_Phi_K1),
         p.value = ifelse(Species %in% improved.fits.species,p.value_Phi_K2,p.value_Phi_K1),
         adj_p.value = p.adjust(p.value,method="BH"))




compare.emp <- compare.emp %>% 
  left_join(compare.emp %>% dplyr::select(Species,estimate_Phi),by=c("Compared.to"="Species"),suffix=c("","_Reference"))
compare.emp <- compare.emp %>%
  mutate(Corr.Diff = estimate_Phi - estimate_Phi_Reference)
compare.emp.non.ref <- compare.emp %>% 
  filter(Corr.Diff != 0)

compare.emp.grouped <- compare.emp.non.ref %>% 
  group_by(Compared.to,Divergence) %>%
  summarize(Mean.Corr.Diff=mean(Corr.Diff),
            Total=n())

corr.diff.plot <- ggplot(compare.emp.non.ref,aes(x=Divergence,y=Corr.Diff)) +
  geom_point() +
  stat_cor(aes(x=Divergence,y=Corr.Diff),method="spearman",label.sep="\n",inherit.aes = F,label.x.npc = "center",label.y = -0.5,cor.coef.name = "rho") +
  theme_cowplot() +
  ylab("Difference in Spearman Correlation") +
  xlab("Divergence time (MYA)") +
  #ggtitle("Correlation between ROC and RNA-Seq\nRelative to Reference Species") +
  theme(aspect.ratio=1)
corr.diff.plot

ggsave2("../Figures/All_species/phi_relative_to_refence_species.pdf",plot=corr.diff.plot,height=5,width=5)

```









```{r fig.width=9}
tree.plot <- ggtree(fungi.tree)

const.tree.plot <- treeTraitPlot(tree.plot,compare.emp,"estimate_Phi_K1","Spearman correlation\nConstMut")
var.tree.plot <- treeTraitPlot(const.tree.plot,compare.emp,"estimate_Phi_K2","Spearman correlation\nVarMut")
var.tree.plot <- treeTraitPlot(var.tree.plot,compare.emp,"Diff","Difference\nVarMut - ConstMut")
var.tree.plot <- var.tree.plot + theme_tree2()
var.tree.plot
cor.tree.plot <- treeTraitPlot(tree.plot,compare.emp,"estimate_Phi","Spearman correlation\nEmp. vs. Pred. Gene Expr.")
cor.tree.plot <- cor.tree.plot + theme_tree2()
cor.tree.plot
ggsave2("../Figures/All_species/constmut_vs_varmut.pdf",plot = var.tree.plot,width=9)
ggsave2("../Figures/All_species/final_phi.pdf",plot = cor.tree.plot,width=9)


```



```{r}
print(nrow(compare.emp %>% filter(estimate_Phi > 0))/327)
print(paste("Number of species for which VarMut model improved model fit by 25%:",length(improved.fits.species)/327))

compare.emp %>% filter(Species %in% improved.fits.species) %>% summarize(median(Diff))
compare.emp %>% filter(!Species %in% improved.fits.species) %>% summarize(median(Diff))
```




