---
title: "Calculate Elongation Waiting Times from tRNA"
author: "Alex Cope"
date: '2023-05-04'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(Biostrings)
library(ape)
library(AnaCoDa)
source("helperFunctions.R")


removeMitoContigs <- function(all.contigs.file,mito.contigs.file,all.input.dir = "/data2/Labella2019/Genomes/fas/",mito.input.dir = "/data2/Labella2019/Genomes/Mito_sequences_removed/",output.dir = "/data2/Labella2019/Genomes/fas_nuc/")
{
  all.contigs <- readDNAStringSet(file.path(all.input.dir,all.contigs.file))
  mito.contigs <- readDNAStringSet(file.path(mito.input.dir,mito.contigs.file))
  mito.id <- names(mito.contigs)
  all.contigs.nuc <- all.contigs[!(names(all.contigs) %in% mito.id)]
  writeXStringSet(all.contigs.nuc,file.path(output.dir,all.contigs.file))
}


gettAI <- function(tgcn,codon.order,sking)
{
  ws <- tAI::get.ws(tgcn,sking=sking)
  tmp <- data.frame(Codon=codon.order,tAI=ws)
  return(tmp)
}

get.ws.local <- function(tRNA, codon.order, sking = 0, s = c(0.0, 0.0, 0.0, 0.0, 0.41, 0.28, 0.9999, 0.68, 0.89), exclude.ctg=F)     # super kingdom: 0-eukaryota, 1-prokaryota
{
  p = 1 - s
  # initialise w vector
  W = NULL  # don't confuse w (lowercase) and W (uppercase)
  
  # obtain absolute adaptiveness values (Ws)
  for (i in seq(1, 61, by=4))
    W = c(W,
          p[1]*tRNA[i]   + p[5]*tRNA[i+1],     # INN -> NNT, NNC, NNA
          p[2]*tRNA[i+1] + p[6]*tRNA[i],       # GNN -> NNT, NNC
          p[3]*tRNA[i+2] + p[7]*tRNA[i],       # TNN -> NNA, NNG
          p[4]*tRNA[i+3] + p[8]*tRNA[i+2])     # CNN -> NNG
  
  # check methionine
  W[36] = p[4]*tRNA[36]
  
  # if bacteria, modify isoleucine ATA codon
  if(sking == 1) W[35] = p[9]
  
  # get rid of stop codons (11, 12, 15) and methionine (36)
  if(exclude.ctg)
  {
    W = W[-c(11,12,15,20,36)] ## drop S-CTG (20) from consideration
    codon.order <- codon.order[-c(11,12,15,20,36)]
  } else {
    W = W[-c(11,12,15,36)]
    codon.order <- codon.order[-c(11,12,15,36)]
  }
  # get ws
  w = W/max(W)
  
  if(sum(w == 0) > 0) {
    ws <- w[w != 0] # zero-less ws
    gm <- exp(sum(log(ws))/length(ws)) # geometric mean
    w[w == 0] = gm # substitute 0-ws by gm
  }
  
  w <- data.frame(Codon = codon.order, tAI = w)
  
  return(w)
}



gettAI.relative.aa <- function(tRNA, codon.order, sking = 0, s = c(0.0, 0.0, 0.0, 0.0, 0.41, 0.28, 0.9999, 0.68, 0.89), exclude.ctg=F)
{

  p = 1 - s

  # initialise w vector
  W = NULL  # don't confuse w (lowercase) and W (uppercase)

  # obtain absolute adaptiveness values (Ws)
  for (i in seq(1, 61, by=4))
    W = c(W,
      p[1]*tRNA[i]   + p[5]*tRNA[i+1],     # INN -> NNT, NNC, NNA
      p[2]*tRNA[i+1] + p[6]*tRNA[i],       # GNN -> NNT, NNC
      p[3]*tRNA[i+2] + p[7]*tRNA[i],       # TNN -> NNA, NNG
      p[4]*tRNA[i+3] + p[8]*tRNA[i+2])     # CNN -> NNG

  # check methionine
  W[36] = p[4]*tRNA[36]
  
  # if bacteria, modify isoleucine ATA codon
  if(sking == 1) W[35] = p[9]

  # get rid of stop codons (11, 12, 15) and methionine (36)
  W = W[-c(11,12,15,36)]

  W <- data.frame(Codon=codon.order,W=W)
  W <- W %>% 
    mutate(AA = unlist(purrr::map(Codon, ~codonToAA(.x))))
  if (exclude.ctg)
  {
    W <- W %>% 
      mutate(AA = ifelse(Codon == "CTG","J",AA),
             W = ifelse(Codon == "CTG",NA,W))
  }
    
  r.wi <- W %>%
        group_by(AA) %>%
        mutate(Order=rank(Codon),
              tAI = 1/W - 1/W[which.max(Order)])
  r.wi <- r.wi %>% 
    ungroup() %>%
    dplyr::select(AA,Codon,Order,W,tAI)
  return(r.wi)
}

```

# Set up color schemes

```{r}
scale_colour_brewer_d <- function(..., palette = "Set3") {
  scale_colour_brewer(..., palette = palette )
}

scale_fill_brewer_d <- function(..., palette = "Set1") {
  scale_fill_brewer(..., palette = palette)
  
}

scale_colour_brewer_c <- function(..., palette = "BuGn") {
  scale_colour_distiller(..., palette = palette )
}

scale_fill_brewer_c <- function(..., palette = "BuGn") {
  scale_fill_distiller(..., palette = palette)
}

options(
  ggplot2.discrete.colour = scale_colour_brewer_d,
  ggplot2.discrete.fill = scale_fill_brewer_d,
  ggplot2.continuous.colour = scale_colour_brewer_c,
  ggplot2.continuous.fill = scale_fill_brewer_c
)
```

# Get new tGCN estimates

```{r}
all.genomes <- readLines("../Data/all_fungi.txt")
ser.genomes <- readLines("../Data/ser_fungi.txt")
fungi.tree <- read.tree("../Data/tree_with_cds_labels.nwk")
fungi.tree <- cleanTree(fungi.tree,all.genomes)

trna.count <- read_tsv("../labella_trna.csv")
trna.count <- trna.count %>%
  dplyr::select(-c(`Species name`,`Major clade`,`Total_tRNA Counts`)) %>%
  dplyr::rename(Species=file_name_id) %>%
  mutate(Species = paste0(Species,".max.cds")) %>%
  filter(Species %in% all.genomes)


# codons <- as.character(
#   reverseComplement(
#     DNAStringSet(
#       str_extract(colnames(trna.count)[2:ncol(trna.count)],"[ACTG]{3}"
#                   ))))
# colnames(trna.count)[2:ncol(trna.count)] <- codons

# trna.count <- trna.count %>% 
#   mutate(CTG=ifelse(is.na(CTG),0,CTG))

trna.count.long <- trna.count %>% 
  pivot_longer(-Species,names_to="Codon", values_to="tGCN")

getNewTGCN <- function(species,old.trna.count)
{
  print(species)
  species.fas <- str_replace(species,".max.cds",".fas")
  trna.scan <- read_tsv(file.path("../Data/new_tGCN/",species.fas),skip = 1,col_types = cols()) %>%
    filter(Codon != "-----") %>%
    mutate(Note = ifelse(is.na(Note),"",Note)) %>%
    filter(Note != "pseudo") %>%
    mutate(Codon = paste(Codon,Type,sep="-")) %>%
    group_by(Codon) %>%
    summarize(tGCN=n()) %>%
    ungroup()
  
  old.tgcn <- old.trna.count %>% 
    filter(Species == species)
  
  new.tgcn <- old.tgcn %>% 
    full_join(trna.scan,by="Codon",suffix=c("_Old","_New")) %>%
    #filter(!Codon %in% c("TAA","TGA","TAG")) %>%
    #mutate(Type = ifelse(is.na(Type),"",Type)) %>%
    mutate(tGCN_New = ifelse(is.na(tGCN_New),0,tGCN_New)) %>%
    filter(Codon != "CTG-Ser" & Codon != "CAT-iMet") %>%
    dplyr::select(Species,Codon,tGCN_New) %>%
    dplyr::rename(tGCN=tGCN_New)

  return(new.tgcn)
}

tgcn.df <- lapply(all.genomes,getNewTGCN,old.trna.count=trna.count.long) %>% 
  bind_rows()
tgcn.df.filt <- tgcn.df %>%
  filter(!str_detect(Codon,"Sup") & !str_detect(Codon,"NNN") & !is.na(Species))
tgcn.df.wide <- tgcn.df.filt %>% 
  pivot_wider(id_cols = Species,names_from = "Codon",values_from="tGCN")
write_csv(tgcn.df.wide,"../Data/cope_trna.csv")
```
```{r}
lab.trna.count <- read_tsv("../Data/labella_trna.tsv")
lab.trna.count <- lab.trna.count %>%
  dplyr::select(-c(`Species name`,`Major clade`,`Total_tRNA Counts`,`CAG-Ala`,`CAG-Ser`)) %>%
  dplyr::rename(Species = file_name_id) %>%
  mutate(Species = paste0(Species,".max.cds")) %>%
  filter(Species %in% all.genomes)
codons <- as.character(
  reverseComplement(
    DNAStringSet(
      str_extract(colnames(lab.trna.count)[2:ncol(lab.trna.count)],"[ACTG]{3}"
                  ))))
colnames(lab.trna.count)[2:ncol(lab.trna.count)] <- codons


```
```{r}
cope.long <- trna.count %>%
 pivot_longer(-Species,names_to="Codon",values_to="tGCN")

lab.long <- lab.trna.count %>%
 pivot_longer(-Species,names_to="Codon",values_to="tGCN")

compare.tgcn <- cope.long %>%
  left_join(lab.long,by=c("Species","Codon"),suffix=c("_Cope","_LaBella")) %>%
  mutate(Diff = tGCN_Cope - tGCN_LaBella)


ggplot(compare.tgcn,aes(x=tGCN_Cope,y=tGCN_LaBella)) +
  geom_point() +
  stat_cor(method="spearman",label.sep="\n") +
  xlab("tGCN per codon\nThis study")+
  ylab("tGCN per codon\nLaBella et a.") +
  geom_abline(intercept=0,slope=1,linetype="dashed") +
  theme_cowplot()



```


```{r}
all.genomes <- readLines("../Data/all_fungi.txt")
ser.genomes <- readLines("../Data/ser_fungi.txt")
fungi.tree <- read.tree("../Data/tree_with_cds_labels.nwk")
fungi.tree <- cleanTree(fungi.tree,all.genomes)

trna.count <- read_csv("../Data/cope_trna.csv")
trna.count <- trna.count %>%
  #dplyr::select(-c(`Species name`,`Major clade`,`Total_tRNA Counts`,`CAG-Ala`,`CAG-Ser`)) %>%
  dplyr::select(-c(`CAG-Ala`,`CAG-Ser`)) %>%
  #dplyr::rename(Species = file_name_id) %>%
  #mutate(Species = paste0(Species,".max.cds")) %>%
  filter(Species %in% all.genomes)


codons <- as.character(
  reverseComplement(
    DNAStringSet(
      str_extract(colnames(trna.count)[2:ncol(trna.count)],"[ACTG]{3}"
                  ))))
colnames(trna.count)[2:ncol(trna.count)] <- codons

trna.count <- trna.count %>% 
  mutate(CTG=ifelse(is.na(CTG),0,CTG))

trna.count.long <- trna.count %>% 
  pivot_longer(-Species,names_to="Codon", values_to="tGCN")

codon.order <- readLines("../Data/dos_reis_tai_codon_order.txt")
codon.order.sense <- codon.order[-c(11,12,15,36)]
trna.count.long <- trna.count.long %>% 
                    group_by(Species) %>% 
                   dplyr::slice(match(codon.order,Codon))
trna.count.long <- trna.count.long %>%
  mutate(Exclude.CTG = ifelse(Species %in% ser.genomes,T,F))


trna.count.long.split <- trna.count.long %>% 
  group_by(Species) %>% 
  group_split()
names(trna.count.long.split) <- unique(trna.count.long$Species)

otai.long <- trna.count.long.split %>%
  purrr::map(~gettAI.relative.aa(.x$tGCN,
                                 codon.order.sense, 
                                 sking = 0, 
                                 exclude.ctg = .x$Exclude.CTG[1])) %>%
  bind_rows(.id="Species")  %>%
  dplyr::rename(otAI = tAI)


tgcn.long <- trna.count.long.split %>%
  purrr::map(~gettAI.relative.aa(.x$tGCN,
                                 codon.order.sense, 
                                 s = c(0,0,0,0,rep(0,5)), 
                                 sking = 0, 
                                 exclude.ctg = .x$Exclude.CTG[1])) %>%
  bind_rows(.id="Species") %>%
  dplyr::rename(tGCN = tAI)

otai.wide <- otai.long %>% 
  pivot_wider(id_cols=Species,names_from=Codon,values_from = otAI)

write_tsv(otai.wide,"../Data/cope_wi_otAI_inverse.tsv")


tgcn.wide <- tgcn.long %>% 
  pivot_wider(id_cols=Species,names_from=Codon,values_from = tGCN)

write_tsv(tgcn.wide,"../Data/cope_wi_tGCN_inverse.tsv")

raw.weights <- otai.long %>% 
  pivot_wider(id_cols=Species,names_from=Codon,values_from = W)

write_tsv(raw.weights,"../Data/cope_unnormalized_wi.tsv")


raw.weights <- tgcn.long %>% 
  pivot_wider(id_cols=Species,names_from=Codon,values_from = W)


write_tsv(raw.weights,"../Data/cope_tGCN_with_wobble.tsv")

```
## Make sure tGCN correlates with AA freq


```{r}
tgcn <- read_tsv("../Data/cope_tgcn_per_species.tsv")
aa.freq <- read_tsv("../Post_analysis/amino_acid_frequency_per_species.tsv")

tgcn.long <- tgcn %>%
  pivot_longer(-Species,names_to="Codon",values_to="tGCN") %>%
  rowwise() %>%
  mutate(AA = AnaCoDa::codonToAA(Codon)) %>%
  ungroup() %>%
  group_by(Species,AA) %>%
  summarize(tGCN = sum(tGCN,na.rm=T))
  

aa.freq.long <- aa.freq %>%
  pivot_longer(-Species,names_to = "AA",values_to="AA_Freq")


tgcn.aa.long <- tgcn.long %>%
  left_join(aa.freq.long,by=c("Species","AA")) %>%
  filter(!AA %in% c("M","W","X","L"))

tgcn.aa.cor <- tgcn.aa.long %>%
  nest(data=c(AA,AA_Freq,tGCN)) %>%
  mutate(corr = purrr::map(data,~cor.test(.x$tGCN,.x$AA_Freq,method = "spearman",use="complete.obs") %>% tidy())) %>%
  dplyr::select(-data) %>%
  unnest(corr)

test<- tgcn.aa.long %>% 
  filter(Species == "candida_albicans.max.cds")

```


```{r fig.height=7}
clades <- read_tsv("../Data/clades.txt")

tgcn.aa.cor.w.clades <- tgcn.aa.cor %>% 
  left_join(clades,by=c("Species"="file_name_id"))

aa.vs.tgcn <- ggplot(tgcn.aa.cor.w.clades,aes(x=Major.clade,y=estimate)) + 
  geom_boxplot() +
  geom_point(aes(color=Major.clade),position="jitter",show.legend = F)+
  ylab("Spearman correlation\ntGCN vs. Amino Acid Frequency") +
  theme_cowplot() +
  xlab("Clade") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 

ggsave2("../Figures/All_species/aa_vs_tgcn.pdf",plot = aa.vs.tgcn)
```



```{r}
tgcn <- read_tsv("../Data/cope_tgcn_per_species.tsv")
tgcn.long <- tgcn %>%
  pivot_longer(-Species,names_to="Codon",values_to="tGCN") %>%
  group_by(Species) %>%
  summarize(tGCN = sum(tGCN,na.rm=T)) %>%
  mutate(Species = str_remove(Species,"(_[0-9]*){0,1}.max.cds"),
         Species = tolower(Species))

genome.quality <- read_csv("../Data/genome_quality.csv")

genome.quality <- genome.quality %>%
  dplyr::rename(Species = old_species_id) %>%
  mutate(Species = tolower(Species)) %>%
  dplyr::select(Species,N50,`BUSCO Complete`)

tgcn.genome.qual <- tgcn.long %>%
  left_join(genome.quality)

ggplot(tgcn.genome.qual,aes(x=N50,y=tGCN)) +
  geom_point() +
  stat_cor(method="spearman",label.sep="\n")+
  theme_cowplot() +
  scale_y_log10() +
  theme(aspect.ratio = 1)





```
