---
title: "Evolution of parameters across species"
author: "Alex Cope"
date: '2023-05-04'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ape)
library(phytools)
library(phylogram)
library(cluster)
library(tidyverse)
library(factoextra)
library(dendextend)
library(RColorBrewer)
library(cowplot)
library(reshape2)
library(ggpubr)
library(ggrepel)
library(broom)
library(AnaCoDa)
library(ggtree)
library(patchwork)
library(scales)
library(ComplexHeatmap)
library(ggmagnify)
source("helperFunctions.R")



compareHClustToPhylo <- function(phylo.dendro,csp.matrix,hclust.method,title="",...)
{
  cormat <- cor(csp.matrix,use="pairwise.complete.obs",method="spearman")
  hc <- hclust(as.dist(1-cormat),method = hclust.method)
  hc.dendro <- as.dendrogram(hc)
  dend.list <- dendlist(phylo.dendro,hc.dendro)
  corr <- cor.dendlist(dend.list,method="cophenetic")
  
  tanglegram(dend.list,highlight_distinct_edges = FALSE, # Turn-off dashed lines
             common_subtrees_color_lines = FALSE, # Turn-off line colors
             common_subtrees_color_branches = TRUE, # Color common branches
             main_left = paste(title,"\nCorrlelation =", round(corr[1,2], 2)),
             tip_labels_cex = 0
  )
}

leg <- function(){
  plot.new()
  par(margin=c(1,1,1,1))
  legend(x = "topright",legend = unique(clade.colors$Clade), fill=unique(clade.colors$Color),ncol=4,cex=1.5,bty='n')
}




```

# Set up color schemes

```{r}
scale_colour_brewer_d <- function(..., palette = "Set3") {
  scale_colour_brewer(..., palette = palette )
}

scale_fill_brewer_d <- function(..., palette = "Set1") {
  scale_fill_brewer(..., palette = palette)
  
}

scale_colour_brewer_c <- function(..., palette = "BuGn") {
  scale_colour_distiller(..., palette = palette )
}

scale_fill_brewer_c <- function(..., palette = "BuGn") {
  scale_fill_distiller(..., palette = palette)
}

options(
  ggplot2.discrete.colour = scale_colour_brewer_d,
  ggplot2.discrete.fill = scale_fill_brewer_d,
  ggplot2.continuous.colour = scale_colour_brewer_c,
  ggplot2.continuous.fill = scale_fill_brewer_c
)
```

# Setup


```{r}
results.dir <- "../Final_runs/Results_k_1/"

all.genomes <- readLines("../Data/all_fungi.txt")
ser.genomes <- readLines("../Data/ser_fungi.txt")
fungi.tree <- read.tree("../Data/tree_with_cds_labels.nwk")
fungi.tree <- cleanTree(fungi.tree,all.genomes)
all.species <- readLines("../Data/all_fungi.txt")

species.dend <- as.dendrogram(fungi.tree)

clades <- read_tsv("../Data/clades.txt")
clades <- clades %>% 
  dplyr::slice(match(fungi.tree$tip.label,file_name_id)) %>%
  dplyr::rename(Species=file_name_id,Clade=Major.clade)


gc.files <- file.path("../Data/GC",all.genomes)
names(gc.files) <- all.genomes
gc.df <- purrr::map(gc.files,read_tsv,col_types = cols()) %>% bind_rows(.id="Species")
gc.df <- gc.df %>% dplyr::rename(GeneID=Gene)

clust.files <- paste0("../Data/Clara_k_2/",all.genomes)
names(clust.files) <- all.genomes
clust.df <- purrr::map(clust.files,read_tsv,col_types = cols()) %>% 
  bind_rows(.id="Species")
clust.df <- clust.df %>% dplyr::rename(GeneID=Gene)

gc.clust.df <- clust.df %>% left_join(gc.df,by=c("Species","GeneID"))

max.cluster <- gc.clust.df  %>% group_by(Species,Cluster) %>% 
               summarize(Median.GC3 = median(GC3)) %>% 
               top_n(1) %>%
               dplyr::rename(Max.Cluster=Cluster) %>%
               dplyr::select(-Median.GC3)

gc.clust.df <- gc.clust.df %>% left_join(max.cluster,by="Species")
gc.clust.df <- gc.clust.df %>% 
              mutate(Cluster = case_when(Cluster == 1 & Max.Cluster == 1 ~ "Higher GC3",
                                         Cluster == 2 & Max.Cluster == 1 ~ "Lower GC3",
                                         Cluster == 1 & Max.Cluster == 2 ~ "Lower GC3",
                                         Cluster == 2 & Max.Cluster == 2 ~ "Higher GC3")
              )



otai <- read_tsv("../Data/cope_wi_otAI_inverse.tsv") %>%
  dplyr::select(-c(TGG)) %>%
  dplyr::slice(match(fungi.tree$tip.label,Species))


tgcn <- read_tsv("../Data/cope_wi_tGCN_inverse.tsv") %>%
  dplyr::select(-c(TGG)) %>%
  dplyr::slice(match(fungi.tree$tip.label,Species))


improved.fits.species <- readLines("../Data/improved_species_by_gene_expression_25_percent_2023_05_03.txt")
compare.emp <- read_csv("../Data/2023-05-04_compare_phi_vs_empirical_k1_vs_k2_roc_fits_all_species.csv")




```



```{r}
deta.matrix.k1 <- read_tsv("../Post_analysis/selection_coefficients_k1.tsv",col_types = cols())
deta.matrix.k2 <- read_tsv("../Post_analysis/selection_coefficients_k2.tsv",col_types = cols())

deta.matrix.k1.rescaled <- read_tsv("../Post_analysis/selection_coefficients_k1_GC_rescaled.tsv",col_types = cols())
deta.matrix.k2.rescaled <- read_tsv("../Post_analysis/selection_coefficients_k2_GC_rescaled.tsv",col_types = cols())

dm.matrix.k1 <- read_tsv("../Post_analysis/mutation_bias_k1.tsv",col_types = cols())
dm.matrix.k2.higher.gc <- read_tsv("../Post_analysis/mutation_bias_k2_higher_GC3.tsv",col_types = cols())
dm.matrix.k2.lower.gc <- read_tsv("../Post_analysis/mutation_bias_k2_lower_GC3.tsv",col_types = cols())

dm.matrix.k1.rescaled <- read_tsv("../Post_analysis/mutation_bias_k1_GC_rescaled.tsv",col_types = cols())
dm.matrix.k2.higher.gc.rescaled <- read_tsv("../Post_analysis/mutation_bias_k2_higher_GC3_GC_rescaled.tsv",col_types = cols())
dm.matrix.k2.lower.gc.rescaled <- read_tsv("../Post_analysis/mutation_bias_k2_lower_GC3_GC_rescaled.tsv",col_types = cols())


```

```{r}
createEmpvsROCdf <- function(target.species,phi,dist.matrix,orthogroups,empirical.data)
{
  dist.from.target <- dist.matrix[,target.species]
  closest.species <- names(dist.from.target[which.min(dist.from.target)])
  
  phi.target <- phi %>% 
    dplyr::select(Orthogroups,!!as.name(target.species)) %>%
    dplyr::rename(Phi = !!as.name(target.species))

  
  ortho.species <- orthogroups %>% 
    dplyr::select(Orthogroups,!!as.name(target.species),!!as.name(closest.species))
  ortho.species <- ortho.species %>% 
    left_join(empirical.data[[closest.species]],
              by=setNames("GeneID",closest.species )
              )
  ortho.species <- ortho.species %>% 
    left_join(phi.target,by="Orthogroups") %>%
    dplyr::select(Orthogroups,Empirical,Phi) %>%
    filter(!is.na(Empirical) & !is.na(Phi)) 
  return(ortho.species)
}

dist.matrix <- cophenetic.phylo(fungi.tree)

species.w.emp <- list.files("../Data/Expression/",full.names = F,pattern=".max.cds")
empirical.data.file <- list.files("../Data/Expression/",full.names = T,include.dirs = F,pattern=".max.cds")
names(empirical.data.file) <- species.w.emp


dist.matrix <- dist.matrix[species.w.emp,]

empirical.data <- purrr::map(empirical.data.file,function(x) {
  read_csv(x,col_types = cols()) %>% 
    mutate(across(contains("tpm_kallisto"),~na_if(., 0))) %>%
    mutate(across(contains("tpm_kallisto"),log)) %>%
    mutate(
      Log.Mean.Expression=rowMeans(dplyr::select(.,contains("tpm_kallisto")),na.rm=T),
      Empirical=exp(Log.Mean.Expression),
      GeneID=str_remove(GeneID,"-mRNA-1")) %>% 
      dplyr::select(-contains("tpm_kallisto"),-Log.Mean.Expression)
    })

phi.df <- read_csv("../Data/phi_shen_etal_ortholog_matrix_2023_05_01.csv",col_types = cols())  %>% 
       column_to_rownames("Orthogroup") %>%
       rename_with(~cleanColumnNames(.x,fungi.tree$tip.label)) %>%
       rownames_to_column("Orthogroups")



orthogroups <- read_tsv("../Data/shen_etal_ortholog_matrix_2021_12_17.tsv",col_types = cols())
orthogroups <- orthogroups %>% 
       column_to_rownames("Orthogroup") %>%
       rename_with(~cleanColumnNames(.x,fungi.tree$tip.label)) %>%
       rownames_to_column("Orthogroups")

names(all.species) <- all.species


roc.emp.df <- lapply(all.species,createEmpvsROCdf,
                      phi = phi.df,
                      dist.matrix = dist.matrix,
                      orthogroups = orthogroups,
                      empirical.data = empirical.data) %>% 
  bind_rows(.id="Species")


write_tsv(roc.emp.df,"../Post_analysis/roc_phi_vs_empirical_for_all_327_yeasts_based_on_shen_etal_2018_one_to_one_orthologs.tsv")



```


### Protein production rates $\phi$

```{r}
getPhi <- function(species,improved.fits.species)
{
  if (species %in% improved.fits.species){
    phi.df <- read_csv(file.path("../Final_runs/Results_k_2_selectionShared/",species,"run_2/Parameter_est/gene_expression.txt"),col_types = cols()) 
  } else{
    phi.df <- read_csv(file.path("../Final_runs/Results_k_1/",species,"run_2/Parameter_est/gene_expression.txt"),col_types = cols()) 
  }
  return(phi.df)
}


species <- list.dirs(file.path("../Final_runs/Results_k_2_selectionShared/"),full.names = F,recursive = F)
names(species) <- species

phi.df <- purrr::map(species,getPhi,improved.fits.species) %>% 
  bind_rows(.id="Species")

phi.summary <- phi.df %>% 
  group_by(Species) %>%
  summarize(Mean.Phi = mean(Mean),
            Sd.Phi = sd(Mean.log10),
            Total = n(),
            )

mean.phi.dist <- ggplot(phi.summary,aes(x=Mean.Phi)) + 
  geom_histogram(fill="white",color="black") +
  theme_cowplot() +
  ggtitle("Distribution of mean gene expression estimate across species")
mean.phi.dist



```


```{r,echo=F}

phi <- read_csv("../Data/phi_shen_etal_ortholog_matrix_2023_05_01.csv")
phi <- phi %>% column_to_rownames("Orthogroup") %>%
       rename_with(~cleanColumnNames(.x,fungi.tree$tip.label))
phi <- phi[,fungi.tree$tip.label]
phi.log <- as.data.frame(phi) %>% 
  mutate(across(where(is.numeric),log)) %>%
  as.matrix()


```

```{r fig.height=8,fig.width=8}
cor.roc.emp <- roc.emp.df %>% 
  group_by(Species) %>%
  nest(data=c(Orthogroups,Empirical,Phi)) %>% 
  mutate(corr.roc.emp = purrr::map(data,~cor.test(.x$Empirical,.x$Phi,method="spearman",use="complete.obs") %>% tidy())
         ) %>% 
  dplyr::select(-data) %>% 
  unnest(starts_with("corr"),names_sep = "_") 

cor.roc.emp <- cor.roc.emp %>% 
  ungroup() %>% 
  mutate(corr.roc.emp_p.value.adjusted = p.adjust(corr.roc.emp_p.value,method = "BH")) %>%
  left_join(clades,by="Species")


roc.vs.emp.box <- ggplot(cor.roc.emp,aes(x=Clade,y=corr.roc.emp_estimate)) + 
  geom_boxplot() +
  geom_point(position="jitter",aes(color=Clade),size=2,show.legend = F) +
  theme_cowplot() + 
  scale_color_brewer(palette="Set3") +
  ylab("Spearman correlation\nPredicted vs. empirical expression") +
  theme(axis.text.x = element_text(angle = 25, vjust = 1, hjust=1),
        axis.title.x = element_blank())

roc.vs.emp.box
```



# Create final parameter matrices dependent on improved species fits 

```{r}
deta.matrix <- deta.matrix.k1
deta.matrix <- deta.matrix %>% 
  column_to_rownames("Species")
deta.matrix <- as.data.frame(t(deta.matrix))

dm.matrix <- dm.matrix.k1
dm.matrix <- dm.matrix %>% 
  column_to_rownames("Species")
dm.matrix <- as.data.frame(t(dm.matrix))


deta.matrix.gc <- deta.matrix.k1.rescaled
deta.matrix.gc <- deta.matrix.gc %>% 
  column_to_rownames("Species")
deta.matrix.gc <- as.data.frame(t(deta.matrix.gc))

dm.matrix.gc <- dm.matrix.k1.rescaled
dm.matrix.gc <- dm.matrix.gc %>% 
  column_to_rownames("Species")
dm.matrix.gc <- as.data.frame(t(dm.matrix.gc))



for (i in improved.fits.species)
{
  deta.matrix <- deta.matrix %>%
    mutate(!!as.name(i) := deta.matrix.k2 %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    )
  deta.matrix.gc <- deta.matrix.gc %>% 
    mutate(!!as.name(i) := deta.matrix.k2.rescaled %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    )
  
  dm.matrix <- dm.matrix %>% 
    mutate(!!as.name(i) := dm.matrix.k2.higher.gc%>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist(),
           !!as.name(paste0(i,"_Lower_GC3")) := dm.matrix.k2.lower.gc %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    ) %>%
      dplyr::rename(!!as.name(paste0(i,"_Higher_GC3")) := !!as.name(i))
  
  dm.matrix.gc <- dm.matrix.gc %>% 
    mutate(!!as.name(i) := dm.matrix.k2.higher.gc.rescaled%>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist(),
           !!as.name(paste0(i,"_Lower_GC3")) := dm.matrix.k2.lower.gc.rescaled %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    ) %>%
      dplyr::rename(!!as.name(paste0(i,"_Higher_GC3")) := !!as.name(i))
}

param.list <- list(deta.matrix,deta.matrix.gc,dm.matrix,dm.matrix.gc) %>%
  purrr::map(~.x %>%
              t() %>% 
              as.data.frame() %>%
              rownames_to_column("Species") %>%
              as_tibble())

deta.matrix <- param.list[[1]]
deta.matrix.gc <- param.list[[2]]
dm.matrix <- param.list[[3]]
dm.matrix.gc <- param.list[[4]]


```

# tGCN across species

Here, we calculate codon-specific weights based on the original tAI.

```{r fig.height=7,fig.width=12}

raw.weights <- read_tsv("../Data/cope_unnormalized_wi.tsv") %>% 
  dplyr::select(-c(TGG)) %>%
  mutate(across(where(is.numeric),log)) %>%
  dplyr::slice(match(fungi.tree$tip.label,Species))


clades <- clades %>%
  dplyr::slice(match(fungi.tree$tip.label,Species))

label.order <- fungi.tree$tip.label


major.clade.names <- sort(unique(unlist(clades[,"Clade"])))

col.map <- brewer.pal(n=length(major.clade.names),name = "Set3")

names(col.map) <- unlist(major.clade.names)
clade_annot <- rowAnnotation(Clade = clades$Clade,
                           col = list(Clade = col.map))

tai.matrix.for.heatmap <- raw.weights %>%
  mutate(across(where(is.numeric), ~na_if(., Inf)), 
         across(where(is.numeric), ~na_if(., -Inf)),
         #across(where(is.numeric), ~na_if(.,0))  
         ) %>% 
  left_join(clades)
tai.matrix.for.heatmap <- tai.matrix.for.heatmap %>%
  dplyr::select(-Clade) %>%
  dplyr::slice(match(fungi.tree$tip.label,Species)) %>%
  column_to_rownames("Species") %>% 
  as.matrix()


col_fun_tai <- circlize::colorRamp2(c(min(tai.matrix.for.heatmap,na.rm=T) - 0.01,0,
                                      max(tai.matrix.for.heatmap,na.rm=T) + 0.01),
                                    c("#e41a1c", "white", "#377eb8")
                                    )

tai.phylo.hm <- Heatmap(tai.matrix.for.heatmap,
        col = col_fun_tai,
        na_col = "black",
        cluster_rows = as.dendrogram(fungi.tree),
        name = "Unnormalized tAI weights",
        left_annotation = clade_annot,
        show_row_names = F)

tai.phylo.hm


cairo_pdf("../Figures/All_species/unnormalized_tai_weights_heatmap_phylogeny.pdf",width=12,height=7,family = "Arial Unicode MS")
draw(tai.phylo.hm)
dev.off()


```





# Compare selection coefficients and waiting times within each species

```{r fig.height=9,fig.width=9, warning = F}
otai <- otai[,colnames(deta.matrix)]
tgcn <- tgcn[,colnames(deta.matrix)]

deta.matrix.long <- deta.matrix %>%
  pivot_longer(-Species,names_to="Codon",values_to="Selection")

tai.matrix.long <- otai %>% 
  pivot_longer(-Species,names_to="Codon",values_to="otAI") %>%
  left_join(tgcn %>% pivot_longer(-Species,names_to="Codon",values_to="tGCN"),by=c("Species","Codon")) %>%
  mutate(otAI = ifelse(is.finite(otAI),otAI,NA),
         tGCN = ifelse(is.finite(tGCN) & tGCN != 0,tGCN,NA))

deta.tai.matrix.long <- deta.matrix.long %>% 
  inner_join(tai.matrix.long,by=c("Species","Codon")) %>%
  left_join(clades,by="Species")

write_tsv(deta.tai.matrix.long,"../Post_analysis/roc_sel_coef_vs_tAI_for_all_327_yeasts.tsv")

cor.df.tai <- deta.tai.matrix.long %>% 
  group_by(Species) %>%
  nest(Codon,Selection,otAI,tGCN) %>% 
  mutate(corr.otai = purrr::map(data,~cor.test(.x$otAI,.x$Selection,method="spearman",use="complete.obs") %>% tidy()),
         corr.tgcn = purrr::map(data,~cor.test(.x$tGCN,.x$Selection,method="spearman",use="complete.obs") %>% tidy())
         
         ) %>% 
  dplyr::select(-data) %>% 
  unnest(starts_with("corr"),names_sep = "_") 

cor.df.tai <- cor.df.tai %>% 
  ungroup() %>% 
  mutate(corr.otai_p.value.adjusted = p.adjust(corr.otai_p.value,method = "BH"),
         corr.tgcn_p.value.adjusted = p.adjust(corr.tgcn_p.value,method = "BH"))

tree.plot <- ggtree(fungi.tree)

tai.phylo <- treeTraitPlot(tree.plot,cor.df.tai,"corr.otai_estimate",significance="corr.otai_p.value",panel.name = "Spearman Correlation\nSelection Coefficients vs. otAI")

total.plot <- tai.phylo + 
  theme_tree2() +
  theme(legend.title = element_text(size=16),
        legend.text = element_text(size=14))
total.plot


sel.vs.tai.box <- ggplot(cor.df.tai,aes(x=Clade,y=corr.otai_estimate)) + 
  geom_boxplot() +
  geom_point(position="jitter",aes(color=Clade),size=2,show.legend = F) +
  theme_cowplot() + 
  scale_color_brewer(palette="Set3") +
  ylab("Spearman correlation\nSelection coefficient vs. relative tAI") +
  theme(axis.text.x = element_text(angle = 25, vjust = 1, hjust=1),
        axis.title.x = element_blank())

sel.vs.tai.box 


```

```{r}
compare.emp.tai <- compare.emp %>% 
  left_join(cor.df.tai,by=c("Species","Clade"))
compare.emp.tai <- compare.emp.tai %>% 
  dplyr::slice(match(fungi.tree$tip.label,Species))


tai.vs.phi.cor <- ggplot(compare.emp.tai,aes(x=estimate_Phi,y=corr.otai_estimate)) +
  geom_point(aes(color=Clade)) +
  stat_cor(method="spearman",label.sep="\n",cor.coef.name = "rho") +
  xlab("Spearman correlation\nPredicted vs. empirical expression") +
  ylab("Spearman correlation\nSelection coefficient vs. relative tAI") +
  geom_hline(yintercept=0,linetype="dashed") +
  geom_vline(xintercept=0,linetype="dashed") +
  theme_cowplot() +
  scale_fill_brewer(palette="Set3") +
  theme(aspect.ratio=1)

tai.vs.phi.cor 

ggsave2("../Figures/All_species/tai_vs_phi_cor.pdf",plot=tai.vs.phi.cor)



write_tsv(compare.emp.tai %>% 
  dplyr::select(Species,estimate_Phi,corr.otai_estimate) %>% 
  dplyr::rename(Corr.Gene.Exp.vs.Phi = estimate_Phi, 
                Corr.tAI.vs.Sel.Coef = corr.otai_estimate),"../Post_analysis/within_species_correlaitons_gene_expression_vs_phi_and_otAI_vs_sel_coeff.tsv")


```


```{r,fig.height=10,fig.width=10}
calb.selection <- deta.tai.matrix.long %>% 
  filter(Species == "candida_albicans.max.cds") %>%
  rowwise() %>%
  mutate(AA = codonToAA(Codon),
         AA.Codon = paste(AA,Codon,sep="-")) %>%
  ungroup() %>%
  mutate(Label=ifelse(otAI > 10,AA.Codon,NA))

calb.sel.vs.tai <- ggplot(calb.selection,aes(x=otAI,y=Selection,label=Label)) +
  geom_point() +
  geom_text(nudge_y=0.1,nudge_x=-300) +
  theme_cowplot() +
  geom_magnify(aes(from = otAI < 2),to = c(1000,4250,-0.5,1.0),axes="xy") +
  stat_cor(aes(x=otAI,y=Selection),method="spearman",label.sep = "\n",label.x=500,label.y=1.75,cor.coef.name = "rho") +
  ylab(expression("Selection coefficient"~Delta*eta)) +
  xlab(expression("Relative "~Delta*"1/tAI")) +
  ggtitle(expression(italic("C. albicans")))
calb.sel.vs.tai

calb.phi <-  roc.emp.df%>% 
  filter(Species == "candida_albicans.max.cds")

calb.phi.vs.emp <- ggplot(calb.phi,aes(x=Empirical,y=Phi)) +
  geom_point() +
  theme_cowplot() +
  scale_x_log10(labels = trans_format("log10", math_format(10^.x))) + 
               scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
  stat_cor(method="spearman",label.sep = "\n",cor.coef.name = "rho") +
  ylab("Scaled predicted expression") +
  xlab("Empirical expression (RNA TPM)") +
  ggtitle(expression(italic("C. albicans")))

     
calb.phi.vs.emp

# layout<-"
# AAEBBBB
# AAEBBBB
# CCEDDDD
# CCEDDDD
# "
# sel.vs.trna <- (calb.phi.vs.emp + theme(axis.title.x = element_text(margin = margin(t = -75, unit = "pt")))) + roc.vs.emp.box  + (calb.sel.vs.tai + theme(axis.title.x = element_text(margin = margin(t = -75, unit = "pt")))) + sel.vs.tai.box + plot_spacer() +
#   plot_annotation(tag_level="A") + 
#   plot_layout(design = layout)

layout<-"
AABBBB
AABBBB
CCDDDD
CCDDDD
"

sel.vs.trna <- (calb.phi.vs.emp + theme(axis.title.x = element_text(margin = margin(t = -75, unit = "pt")))) + roc.vs.emp.box  + (calb.sel.vs.tai + theme(axis.title.x = element_text(margin = margin(t = -75, unit = "pt")))) + sel.vs.tai.box +
  plot_annotation(tag_level="A") + 
  plot_layout(design = layout)

sel.vs.trna

savePlot("../Figures/All_species/intraspecies_translational_selection.pdf",plot = sel.vs.trna,height=10,width=12)



```



```{r echo = F, fig.width=9,fig.height=9,warning=F}
plotSelectionAndTAI <- function(codon.df,tree)
{
  codon.df <- codon.df %>% 
    dplyr::slice(match(tree$tip.label,Species))
  tree.plot <- ggtree(tree)

  tree.plot <- treeTraitPlot(tree.plot,codon.df,"otAI",significance=NULL,panel.name = "tAI")
  tree.plot <- treeTraitPlot(tree.plot,codon.df,"Selection",significance=NULL,panel.name = "Selection Coefficient")
  total.plot <- tree.plot + 
    theme_tree2() +
    theme(legend.title = element_text(size=16),legend.text = element_text(size=14)) +
    ggtitle(unique(codon.df$Codon))
  total.plot
}

deta.tai.matrix.long.split <- deta.tai.matrix.long %>% 
  group_by(Codon) %>%
  group_split()

lapply(deta.tai.matrix.long.split,plotSelectionAndTAI,tree=fungi.tree)
```



```{r echo = F,warning=F,fig.width=16,fig.height=16}

codon.selection.long <- deta.tai.matrix.long %>%
  pivot_longer(c(Selection,otAI),names_to="Parameter",values_to="Estimate")

codon.dist <- ggplot(codon.selection.long,aes(x=Estimate,fill=Parameter)) +
  geom_histogram(position = "identity",color="black",alpha=0.5) +
  geom_vline(xintercept = 0,linetype="solid",color="black") +
  theme_cowplot() +
  theme(aspect.ratio=1) +
  ggtitle("Distribution of selection coefficients and tAI") +
  ylab("Count") +
  facet_wrap(~Codon,scales = "free")
codon.dist

savePlot("../Figures/All_species/S04_selection_and_tai_distributions.pdf",plot=codon.dist,height=16,width=16)

```


