---
title: "Compare Selection and Mutation to Phylogeny"
author: "Alexander Cope"
date: "1/25/2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ape)
library(phytools)
library(phylogram)
library(cluster)
library(tidyverse)
library(factoextra)
library(dendextend)
library(RColorBrewer)
library(cowplot)
library(reshape2)
library(geiger)
library(ggpubr)
library(ggrepel)
library(broom)
library(AnaCoDa)
library(ggtree)
library(patchwork)
library(scales)
library(ComplexHeatmap)
library(ggnewscale)
source("helperFunctions.R")

cleanSpeciesLabelsForPlots <- function(x)
{
   y <- x %>% str_remove(".max.cds") %>% 
          str_extract("[a-z]+_[a-z]+") %>%
          str_split("_") %>% 
          lapply(function(x){paste0(toupper(substring(x[1],1,1)),". ",x[2])})
          #lapply(function(x){paste(str_to_sentence(x[1]),x[2])})
  unlist(y)
}


plotCodonRank <- function(df)
{
  opt.freq <- df %>%
    ungroup() %>%
    group_by(Codon,AA,Codon.Rank) %>%
    summarize(Opt.Freq = n()/length(unique(df$Species))) %>%
    mutate(Codon.Rank = ifelse(Codon.Rank == 1,"1 (Favored)",as.character(Codon.Rank)))

  opt.freq.bar <- ggplot(opt.freq,aes(x=Codon,y=Opt.Freq,fill=Codon.Rank)) +
    geom_bar(stat="identity",position = position_dodge(),color="black") + 
    theme_cowplot() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    labs(fill = "Codon Rank") +
    scale_fill_viridis_d() +
    ylab("Frequency assigned codon rank across all species")+ 
    xlab("Codon")
  
  opt.freq.bar <- opt.freq.bar + facet_wrap(~AA,scales="free_x")
  return(opt.freq.bar)
}

```

# Set up color schemes

```{r}
scale_colour_brewer_d <- function(..., palette = "Set3") {
  scale_colour_brewer(..., palette = palette )
}

scale_fill_brewer_d <- function(..., palette = "Set1") {
  scale_fill_brewer(..., palette = palette)
  
}

scale_colour_brewer_c <- function(..., palette = "BuGn") {
  scale_colour_distiller(..., palette = palette )
}

scale_fill_brewer_c <- function(..., palette = "BuGn") {
  scale_fill_distiller(..., palette = palette)
}

options(
  ggplot2.discrete.colour = scale_colour_brewer_d,
  ggplot2.discrete.fill = scale_fill_brewer_d,
  ggplot2.continuous.colour = scale_colour_brewer_c,
  ggplot2.continuous.fill = scale_fill_brewer_c
)
```




# Read in relevant data

## Phylogeny


```{r}
all.genomes <- readLines("../Data/all_fungi.txt")
ser.genomes <- readLines("../Data/ser_fungi.txt")
fungi.tree <- read.tree("../Data/tree_with_cds_labels.nwk")
fungi.tree <- cleanTree(fungi.tree,all.genomes)

clades <- read_tsv("../Data/clades.txt")
clades <- clades %>%
  dplyr::rename(Species=file_name_id,Clade=Major.clade) %>%
  filter(Species %in% fungi.tree$tip.label)

gc.cont <- read_tsv("../Data/gc_content_by_genome.tsv") %>%
           dplyr::rename(Species=Genome) %>%
           dplyr::slice(match(fungi.tree$tip.label,Species)) %>%
           left_join(clades,by="Species")

```


## Selection coefficients and mutation bias estimates across 327 yeasts
```{r}
deta.matrix.k1 <- read_tsv("../Post_analysis/selection_coefficients_k1.tsv",col_types = cols())
deta.matrix.k2 <- read_tsv("../Post_analysis/selection_coefficients_k2.tsv",col_types = cols())
sd.deta.matrix.k1 <- read_tsv("../Post_analysis/std_selection_coefficients_k1.tsv",col_types = cols())
sd.deta.matrix.k2 <- read_tsv("../Post_analysis/std_selection_coefficients_k2.tsv",col_types = cols())


deta.matrix.k1.rescaled <- read_tsv("../Post_analysis/selection_coefficients_k1_GC_rescaled.tsv",col_types = cols())
deta.matrix.k2.rescaled <- read_tsv("../Post_analysis/selection_coefficients_k2_GC_rescaled.tsv",col_types = cols())

dm.matrix.k1 <- read_tsv("../Post_analysis/mutation_bias_k1.tsv",col_types = cols())
dm.matrix.k2.higher.gc <- read_tsv("../Post_analysis/mutation_bias_k2_higher_GC3.tsv",col_types = cols())
dm.matrix.k2.lower.gc <- read_tsv("../Post_analysis/mutation_bias_k2_lower_GC3.tsv",col_types = cols())
sd.dm.matrix.k1 <- read_tsv("../Post_analysis/std_mutation_bias_k1.tsv",col_types = cols())
sd.dm.matrix.k2.higher.gc <- read_tsv("../Post_analysis/std_mutation_bias_k2_higher_GC3.tsv",col_types = cols())
sd.dm.matrix.k2.lower.gc <- read_tsv("../Post_analysis/std_mutation_bias_k2_lower_GC3.tsv",col_types = cols())



dm.matrix.k1.rescaled <- read_tsv("../Post_analysis/mutation_bias_k1_GC_rescaled.tsv",col_types = cols())
dm.matrix.k2.higher.gc.rescaled <- read_tsv("../Post_analysis/mutation_bias_k2_higher_GC3_GC_rescaled.tsv",col_types = cols())
dm.matrix.k2.lower.gc.rescaled <- read_tsv("../Post_analysis/mutation_bias_k2_lower_GC3_GC_rescaled.tsv",col_types = cols())
```

## Species better fit by VarMut model than ConstMut model based on comparisons with empirical gene expression
```{r}
improved.fits.species <- readLines("../Data/improved_species_by_gene_expression_25_percent_2023_05_03.txt")
```

## Create final parameter matrices dependent on improved species fits 

```{r}

deta.matrix <- deta.matrix.k1
deta.matrix <- deta.matrix %>% 
  column_to_rownames("Species")
deta.matrix <- as.data.frame(t(deta.matrix))

sd.deta.matrix <- sd.deta.matrix.k1
sd.deta.matrix <- sd.deta.matrix %>% 
  column_to_rownames("Species")
sd.deta.matrix <- as.data.frame(t(sd.deta.matrix))


dm.matrix <- dm.matrix.k1
dm.matrix <- dm.matrix %>% 
  column_to_rownames("Species")
dm.matrix <- as.data.frame(t(dm.matrix))

sd.dm.matrix <- sd.dm.matrix.k1
sd.dm.matrix <- sd.dm.matrix %>% 
  column_to_rownames("Species")
sd.dm.matrix <- as.data.frame(t(sd.dm.matrix))

deta.matrix.gc <- deta.matrix.k1.rescaled
deta.matrix.gc <- deta.matrix.gc %>% 
  column_to_rownames("Species")
deta.matrix.gc <- as.data.frame(t(deta.matrix.gc))

dm.matrix.gc <- dm.matrix.k1.rescaled
dm.matrix.gc <- dm.matrix.gc %>% 
  column_to_rownames("Species")
dm.matrix.gc <- as.data.frame(t(dm.matrix.gc))



for (i in improved.fits.species)
{
  deta.matrix <- deta.matrix %>%
    mutate(!!as.name(i) := deta.matrix.k2 %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    )
  
  sd.deta.matrix <- sd.deta.matrix %>%
    mutate(!!as.name(i) := sd.deta.matrix.k2 %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    )
  
  deta.matrix.gc <- deta.matrix.gc %>% 
    mutate(!!as.name(i) := deta.matrix.k2.rescaled %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    )
  
  dm.matrix <- dm.matrix %>% 
    mutate(!!as.name(i) := dm.matrix.k2.higher.gc%>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist(),
           !!as.name(paste0(i,"_Lower_GC3")) := dm.matrix.k2.lower.gc %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    ) %>%
      dplyr::rename(!!as.name(paste0(i,"_Higher_GC3")) := !!as.name(i))
  
  sd.dm.matrix <- sd.dm.matrix %>% 
    mutate(!!as.name(i) := sd.dm.matrix.k2.higher.gc%>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist(),
           !!as.name(paste0(i,"_Lower_GC3")) := sd.dm.matrix.k2.lower.gc %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    ) %>%
      dplyr::rename(!!as.name(paste0(i,"_Higher_GC3")) := !!as.name(i))
  
  dm.matrix.gc <- dm.matrix.gc %>% 
    mutate(!!as.name(i) := dm.matrix.k2.higher.gc.rescaled%>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist(),
           !!as.name(paste0(i,"_Lower_GC3")) := dm.matrix.k2.lower.gc.rescaled %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    ) %>%
      dplyr::rename(!!as.name(paste0(i,"_Higher_GC3")) := !!as.name(i))
}

param.list <- list(deta.matrix,deta.matrix.gc,dm.matrix,dm.matrix.gc,sd.deta.matrix,sd.dm.matrix) %>%
  purrr::map(~.x %>%
              t() %>% 
              as.data.frame() %>%
              rownames_to_column("Species") %>%
              as_tibble())

deta.matrix <- param.list[[1]]
deta.matrix.gc <- param.list[[2]]
dm.matrix <- param.list[[3]]
dm.matrix.gc <- param.list[[4]]
sd.deta.matrix <- param.list[[5]]
sd.dm.matrix <- param.list[[6]]


```

## Selection coefficients vs tAI for all species

```{r}
deta.vs.wi <- read_tsv("../Post_analysis/roc_sel_coef_vs_tAI_for_all_327_yeasts.tsv") 
```


## ROC $\phi$ vs. empirical gene expression for all species

```{r}
phi.vs.emp <- read_tsv("../Post_analysis/roc_phi_vs_empirical_for_all_327_yeasts_based_on_shen_etal_2018_one_to_one_orthologs.tsv") 
```

## Selection coefficient vs. tAI and ROC $\phi$ vs. empirical gene expression correlations

```{r}

corr.df <- read_tsv("../Post_analysis/within_species_correlaitons_gene_expression_vs_phi_and_tAI_vs_sel_coeff.tsv")

```

## Codon rankings by selection and mutation bias

### Natural selection

```{r}
deta.matrix.long <- deta.matrix %>%
  pivot_longer(-Species,names_to="Codon",values_to="Selection")

ref.codons <- unlist(lapply(AnaCoDa::aminoAcids(),function(a){
  if(a %in% c("M","W","X")) return(NA)
  codon.vec<-AnaCoDa::AAToCodon(a)
  return(codon.vec[length(codon.vec)])
}))
ref.codons <- ref.codons[which(!is.na(ref.codons))]

ref.df <- data.frame(Species = rep(unique(deta.matrix.long$Species),length(ref.codons)),Codon=ref.codons,Selection=0)

deta.matrix.long <- list(deta.matrix.long,ref.df) %>%
  bind_rows()

deta.matrix.long <- deta.matrix.long %>% 
  rowwise() %>% 
  mutate(AA = AnaCoDa::codonToAA(Codon),
         Codon.AA = paste(AA,Codon,sep="-")) %>%
  ungroup() %>% 
  group_by(Species,AA) %>% 
  mutate(Codon.Rank = rank(Selection)) 


```

```{r fig.width=12,fig.height=12}

addCladeLabels <- function(codon.rank.phylo.heatmap,clades)
{
  codon.rank.phylo.heatmap.2 <- codon.rank.phylo.heatmap + new_scale_fill()
  codon.rank.phylo.heatmap.2 <- gheatmap(codon.rank.phylo.heatmap.2, clades,
                           offset=1000, 
                           width=0.5,
                           color=NA,
                           legend_title = "Clade",
                           colnames_offset_y = -5,
                           colnames_angle = 0,
                           font.size = 0) +
    scale_fill_brewer(palette = "Set3") +
    labs(fill="Clade")
  codon.rank.phylo.heatmap.2 <- codon.rank.phylo.heatmap.2 + 
    theme(legend.text = element_text(size=10),
          legend.title = element_text(size=14),
          legend.margin=margin(t = 0, unit='cm'),
          plot.margin=margin(0,0,0,0, unit = 'cm'),
          plot.title = element_text(hjust=0.5,size=24),#,vjust=-5),
          axis.title = element_blank(),
          axis.text = element_blank(),
          axis.ticks = element_blank()
          )
  return(codon.rank.phylo.heatmap.2)
}

plotChangeInFavoredCodon <- function(codon.ranks,clades)
{
  two.codon.ranks.ag <- codon.ranks[,c("E","K","Q")]
  two.codon.ranks.ct <- codon.ranks[,c("C","D","F","H","N","Y","Z")]
  four.codon.ranks <- codon.ranks[,c("A","G","P","S","T","V")]
  
  p <- ggtree(fungi.tree,layout = "circular")
  
  codon.compare.ag <- gheatmap(p, two.codon.ranks.ag,
                         offset=10, 
                         width=2,
                         color=NA,
                         colnames_offset_y = -5,
                         colnames_level = c("E","K","Q"),
                         font.size = 3) +
    scale_fill_manual(breaks=c("A","G","C","T"),values=c("#e41a1c","#377eb8","#4daf4a","#984ea3"),name="Third nucleotide",drop=F) + 
    labs(fill="none") + 
    scale_y_continuous(expand=c(0, 2))
  
  codon.compare.ct <- gheatmap(p, two.codon.ranks.ct,
                         offset=10, 
                         width=2.20,
                         color=NA,
                         colnames_offset_y = -5,
                         colnames_level = c("C","D","F","H","N","Y","Z"),
                         font.size = 3) +
      scale_fill_manual(breaks=c("A","G","C","T"),values=c("#e41a1c","#377eb8","#4daf4a","#984ea3"),name="Third nucleotide",drop=F) +
    labs(fill="none") + 
    scale_y_continuous(expand=c(0, 2))
      
  
  codon.compare.4 <- gheatmap(p, four.codon.ranks,
                         offset=10, 
                         width=2.20,
                         color=NA,
                         colnames_offset_y = -5,
                         colnames_level = c("A","G","P","S","T","V"),
                         font.size = 3) +
      scale_fill_manual(breaks=c("A","G","C","T"),values=c("#e41a1c","#377eb8","#4daf4a","#984ea3"),name="Third nucleotide",drop=F) + 
      scale_y_continuous(expand=c(0, 2))
  
  codon.compare.ag <- addCladeLabels(codon.compare.ag,clades) + theme(legend.position = "none")
  codon.compare.ct <- addCladeLabels(codon.compare.ct,clades) + theme(legend.position = "none",plot.title = element_blank())
  codon.compare.4 <- addCladeLabels(codon.compare.4,clades) + theme(plot.title = element_blank())
  
  codon.compare.aa <- (codon.compare.ag / codon.compare.ct / codon.compare.4) + plot_layout(guides="collect")
  
  return(codon.compare.aa)
}


codon.sel.ranks <- deta.matrix.long %>% 
  filter(AA %in% c("K","Q","E","C","D","F","H","N","Y","Z","A","G","P","S","T","V") & Codon.Rank == 1) %>%
  ungroup() %>%
  dplyr::select(Species,AA,Codon) %>%
  separate(Codon,into=c("First","Second","Third"),sep = c(1,2)) %>%
  mutate(Third = factor(Third,levels=c("A","G","C","T"))) %>%
  pivot_wider(id_cols = c("Species"),names_from = "AA",values_from="Third") %>%
  column_to_rownames("Species")

clades.mat <- clades %>% 
  dplyr::slice(match(fungi.tree$tip.label,Species)) %>%
  column_to_rownames("Species")



codon.sel.compare <- (plotChangeInFavoredCodon(codon.sel.ranks,clades.mat)) & ggtitle("Selection coefficients")
codon.sel.compare

ggsave2("../Figures/All_species/favored_codons_selection.pdf",plot = codon.sel.compare,width=16,height=16)
```





```{r fig.height= 10}
sel.freq <- plotCodonRank(deta.matrix.long)
sel.freq
ggsave2("../Figures/All_species/frequency_codon_rank_selection.pdf",sel.freq,width=8,height=8)
```


### Mutation bias

```{r}
dm.matrix.long <- dm.matrix %>%
  pivot_longer(-Species,names_to="Codon",values_to="Mutation")


ref.codons <- unlist(lapply(AnaCoDa::aminoAcids(),function(a){
  if(a %in% c("M","W","X")) return(NA)
  codon.vec<-AnaCoDa::AAToCodon(a)
  return(codon.vec[length(codon.vec)])
}))
ref.codons <- ref.codons[which(!is.na(ref.codons))]

ref.df <- data.frame(Species = rep(unique(dm.matrix.long$Species),length(ref.codons)),Codon=ref.codons,Mutation=0)

dm.matrix.long <- list(dm.matrix.long,ref.df) %>%
  bind_rows()

dm.matrix.long.constmut <- dm.matrix.long %>%
  filter(!str_detect(Species,"GC3")) %>% 
  rowwise() %>% 
  mutate(AA = AnaCoDa::codonToAA(Codon)) %>%
  ungroup() %>% 
  group_by(Species,AA) %>% 
  mutate(Codon.Rank = rank(Mutation)) 

dm.matrix.long.lower.gc3 <- dm.matrix.long %>%
  filter(str_detect(Species,"Lower_GC3")) %>% 
  rowwise() %>% 
  mutate(AA = AnaCoDa::codonToAA(Codon)) %>%
  ungroup() %>% 
  group_by(Species,AA) %>% 
  mutate(Codon.Rank = rank(Mutation)) 

dm.matrix.long.higher.gc3 <- dm.matrix.long %>%
  filter(str_detect(Species,"Higher_GC3")) %>% 
  rowwise() %>% 
  mutate(AA = AnaCoDa::codonToAA(Codon)) %>%
  ungroup() %>% 
  group_by(Species,AA) %>% 
  mutate(Codon.Rank = rank(Mutation)) 


```


```{r fig.height= 10}
mut.const.freq <- plotCodonRank(dm.matrix.long.constmut)
mut.const.freq

mut.var.lower.gc3freq <- plotCodonRank(dm.matrix.long.lower.gc3)
mut.var.lower.gc3freq

mut.var.higher.gc3freq <- plotCodonRank(dm.matrix.long.higher.gc3)
mut.var.higher.gc3freq

ggsave2("../Figures/All_species/frequency_codon_rank_mutation_constmut.pdf",mut.const.freq,width=8,height=8)
ggsave2("../Figures/All_species/frequency_codon_rank_mutation_lower_gc3.pdf",mut.var.lower.gc3freq,width=8,height=8)
ggsave2("../Figures/All_species/frequency_codon_rank_mutation_higher_gc3.pdf",mut.var.higher.gc3freq,width=8,height=8)
```

```{r}

codon.mut.ranks <- dm.matrix.long.constmut %>% 
  filter(AA %in% c("K","Q","E","C","D","F","H","N","Y","Z","A","G","P","S","T","V") & Codon.Rank == 1) %>%
  ungroup() %>%
  dplyr::select(Species,AA,Codon) %>%
  mutate(Species = str_remove(Species,"_Higher_GC3")) %>%
  separate(Codon,into=c("First","Second","Third"),sep = c(1,2)) %>%
  pivot_wider(id_cols = c("Species"),names_from = "AA",values_from="Third") %>%
  column_to_rownames("Species")


codon.constmut.compare <- (plotChangeInFavoredCodon(codon.mut.ranks,clades.mat)) & ggtitle("Mutation bias\nConstMut")
codon.constmut.compare


ggsave2("../Figures/All_species/favored_codons_constmut.pdf",plot = codon.constmut.compare,width=11,height=8)


```

```{r fig.width = 10,fig.height=12}


sel.mut.codon <- ((codon.sel.compare + theme(legend.position = "none")) | codon.constmut.compare)
sel.mut.codon
ggsave2("../Figures/All_species/codon_favored_across_species.pdf",plot=sel.mut.codon,width=12,height = 13)

```



```{r echo = F, fig.width=9,fig.height=9}

tree.plot <- ggtree(fungi.tree)

gc.phylo <- treeTraitPlot(tree.plot,
                          gc.cont,
                          "GC",
                          significance=NULL,
                          panel.name = "GC Content",
                          xlim=c(0,1))

total.plot <- gc.phylo + 
  theme_tree2() +
  theme(legend.title = element_text(size=16),legend.text = element_text(size=14))
  #ggtitle("GC frequency across Saccharomycotina")
total.plot

gc.box <- ggplot(gc.cont,aes(x=Clade,y=GC*100)) + 
  geom_boxplot() +
  geom_point(aes(fill=Clade),shape=21,position="jitter",show.legend = F) + 
  theme_cowplot() + 
  ylab("Genome-wide GC%") +
  scale_fill_brewer(palette="Set3") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

gc.box

ggsave2("../Figures/All_species/gc_content.pdf",gc.box)

```

### Predicting GC% from mutation biases

```{r}
extractLastNucleotide <- function(codon) {
  substr(codon, 3, 3)
}

calculateExpectedGC <- function(aa,dm.matrix)
{
  codons <- AAToCodon(aa,focal = T)
  
  ## This function seems to work, but spits out this warning:
  ## Warning in if (old_name %in% codon.columns) { :
  ##   the condition has length > 1 and only the first element will be used
  # mapNewNames <- function(old_name) {
  #   if (old_name %in% codons) {
  #     extractLastNucleotide(old_name)
  #   } else {
  #     old_name
  #   }
  # }
  mapNewNames <- function(old_name) {
      extractLastNucleotide(old_name)
  }
  
  test <- dm.matrix[,c("Species",codons)] %>% 
    rename_with(~mapNewNames(.), .cols = all_of(codons)) 
  test <- test %>%
    mutate(`T` = 0,
           across(c(A,C,G,`T`), ~.x * -1),
           across(c(A,C,G,`T`),exp)) %>% 
    rowwise() %>% 
    mutate(Denom = sum(c(A,C,G,`T`))) %>% 
    ungroup() %>% 
    mutate(across(c(A,C,G,`T`),~.x/Denom)) %>%
    rowwise() %>%
    mutate(!!as.name(paste0("Expected.GC",".",aa)) := sum(c(C,G))) %>%
    dplyr::select(Species,starts_with("Expected"))
  return(test)
}

```




```{r}
four.codon.aa <- c("A","G","P","S","T","V")


expected.gc.df <- lapply(four.codon.aa,calculateExpectedGC,dm.matrix=dm.matrix) %>%
  purrr::reduce(left_join,by="Species") %>%
  mutate(Mean.Expected.GC = rowMeans(across(starts_with("Expected"))),
         Min.Expected.GC = min(across(starts_with("Expected"))),
         Max.Expected.GC = max(across(starts_with("Expected")))) %>%
  rowwise() %>%
  mutate(Median.Expected.GC = median(c_across(starts_with("Expected")))) %>%
  ungroup() %>%
  separate(Species,into=c("Species","GC3.Cluster"),sep="_(?=(Lower|Higher))") %>%
  mutate(GC3.Cluster = case_when(
    is.na(GC3.Cluster)~"ConstMut",
    GC3.Cluster == "Lower_GC3"~"Lower GC3% Set",
    GC3.Cluster == "Higher_GC3"~"Higher GC3% Set")
  ) %>%
  mutate(GC3.Cluster = factor(GC3.Cluster,levels=c("ConstMut","Lower GC3% Set","Higher GC3% Set")))


obs.exp.gc  <- expected.gc.df %>% 
  left_join(gc.cont,by="Species") %>%
  mutate(Deviation = across(starts_with("Expected"),~abs(.x - GC))) %>%
  mutate(K2 = ifelse(Species %in% improved.fits.species,"Yes","No"))
```

```{r}

BinomialTest.Num.Species.Lower.Expected.GC <- function(obs.exp.gc,model.fit=c("ConstMut"))
{
  num.species.less <- obs.exp.gc %>% 
  filter(GC3.Cluster %in% model.fit & Mean.Expected.GC < GC) %>% 
  nrow()

  num.species <- obs.exp.gc %>% 
    filter(GC3.Cluster %in% model.fit) %>% 
    nrow()
  
  binom.test(num.species.less,num.species)
}

BinomialTest.Num.Species.Higher.Expected.GC <- function(obs.exp.gc,model.fit=c("ConstMut"))
{
  num.species.less <- obs.exp.gc %>% 
  filter(GC3.Cluster %in% model.fit & Mean.Expected.GC > GC) %>% 
  nrow()

  num.species <- obs.exp.gc %>% 
    filter(GC3.Cluster %in% model.fit) %>% 
    nrow()
  
  binom.test(num.species.less,num.species)
}


BinomialTest.Num.Species.Consistent.Expected.GC <- function(obs.exp.gc,model.fit=c("ConstMut"))
{
  num.species.less <- obs.exp.gc %>% 
  filter(GC3.Cluster %in% model.fit & Min.Expected.GC < GC & Max.Expected.GC > GC) %>% 
  nrow()

  num.species <- obs.exp.gc %>% 
    filter(GC3.Cluster %in% model.fit) %>% 
    nrow()
  
  binom.test(num.species.less,num.species)
}

BinomialTest.Num.Species.Consistent.Expected.GC(obs.exp.gc,"ConstMut")
BinomialTest.Num.Species.Consistent.Expected.GC(obs.exp.gc,"Lower GC3% Set")
BinomialTest.Num.Species.Consistent.Expected.GC(obs.exp.gc,"ConstMut")
BinomialTest.Num.Species.Consistent.Expected.GC(obs.exp.gc,c("ConstMut","Lower GC3% Set"))
BinomialTest.Num.Species.Lower.Expected.GC(obs.exp.gc,c("ConstMut","Lower GC3% Set"))
BinomialTest.Num.Species.Higher.Expected.GC(obs.exp.gc,"Higher GC3% Set")

```


```{r fig.width=8}
obs.exp.gc.p <- ggplot(obs.exp.gc,aes(x=Mean.Expected.GC,y=GC)) +
  geom_point() +
  geom_errorbarh(aes(xmin = Min.Expected.GC, xmax = Max.Expected.GC),alpha=0.5) +
  geom_abline(intercept=0,slope=1,linetype="dashed") +
  theme_cowplot() +
  xlab("Expected GC%") +
  ylab("Observed GC%") +
  xlim(0,1) +
  ylim(0,1) +
  theme(aspect.ratio = 1) +
  stat_cor(method = "spearman",cor.coef.name = "rho",label.sep = "\n") +
  facet_wrap(~GC3.Cluster)
obs.exp.gc.p

ggsave2("../Figures/All_species/obs_vs_exp_gc_content.pdf",width=8,height=4)

```







## Setting up for plotting heatmaps to examine variation in parameters 

```{r}
#fungi.tree.dm <- fungi.tree
fungi.tree.lower.gc3 <- fungi.tree
fungi.tree.higher.gc3 <- fungi.tree
for (i in improved.fits.species)
{
  #fungi.tree.dm <- bind.tip(fungi.tree.dm,tip.label=paste0(i,"_Higher_GC3"),
  #                            where = which(fungi.tree.dm$tip.label == i))
  #fungi.tree.dm$tip.label[which(fungi.tree.dm$tip.label == i)] <- paste0(i,"_Lower_GC3")
  fungi.tree.lower.gc3$tip.label[which(fungi.tree.lower.gc3$tip.label == i)] <- paste0(i,"_Lower_GC3")
  fungi.tree.higher.gc3$tip.label[which(fungi.tree.higher.gc3$tip.label == i)] <- paste0(i,"_Higher_GC3")
}

```

```{r}


deta.matrix.for.heatmap <- deta.matrix
deta.matrix.for.heatmap <- deta.matrix.for.heatmap %>%
  dplyr::slice(match(fungi.tree$tip.label,Species)) %>%
  column_to_rownames("Species") %>% 
  as.matrix()


dm.matrix.for.heatmap <- dm.matrix
dm.matrix.for.heatmap <- dm.matrix.for.heatmap %>%
  dplyr::slice(match(fungi.tree.lower.gc3$tip.label,Species)) %>%
  mutate(Species = str_remove(Species,"_Lower_GC3")) %>%
  column_to_rownames("Species") %>% 
  as.matrix()


aa.codon <- sapply(colnames(deta.matrix.for.heatmap),codonToAA)
aa.codon <- paste(aa.codon,names(aa.codon),sep="-")

colnames(deta.matrix.for.heatmap) <- aa.codon
colnames(dm.matrix.for.heatmap) <- aa.codon


clades <- clades %>%
  dplyr::slice(match(fungi.tree$tip.label,Species))

label.order <- fungi.tree$tip.label


major.clade.names <- sort(unique(unlist(clades[,"Clade"])))

col.map <- brewer.pal(n=length(major.clade.names),name = "Set3")

names(col.map) <- unlist(major.clade.names)
clade_annot <- rowAnnotation(Clade = clades$Clade,
                           col = list(Clade = col.map))

col_fun_phi <- circlize::colorRamp2(c(min(corr.df$Corr.Gene.Exp.vs.Phi) - 0.01,
                                      -0.25,
                                      0,
                                      0.25,
                                      0.5,
                                      max(corr.df$Corr.Gene.Exp.vs.Phi) + 0.01),
                                    #c("purple", "white", "gold")
                                    brewer.pal(6,"Blues")
                                    )

col_fun_deta <- circlize::colorRamp2(c(min(deta.matrix.for.heatmap,na.rm=T) - 0.01,0,
                                      max(deta.matrix.for.heatmap,na.rm=T) + 0.01),
                                    c("#377eb8", "white", "#e41a1c")
                                    )
col_fun_dm <- circlize::colorRamp2(c(min(dm.matrix.for.heatmap,na.rm=T) - 0.01,0,
                                      max(dm.matrix.for.heatmap,na.rm=T) + 0.01),
                                    c("#377eb8", "white", "#e41a1c")
                                    )




corr.df.ordered <- corr.df %>% 
  dplyr::slice(match(label.order, Species)) 
improved.ordered <- ifelse(label.order %in% improved.fits.species,"Yes","No")

row_annot <- rowAnnotation(`Emp. vs. Pred. Gene Expr.` = corr.df.ordered$Corr.Gene.Exp.vs.Phi,
                               `VarMut fit` = improved.ordered,
                           col = list(`Emp. vs. Pred. Gene Expr.` = col_fun_phi,
                                      `VarMut fit` = c("Yes" = "black","No"="white")
                           ),
                          annotation_label = c("Emp. vs. Pred. Gene Expr.",
                                               "VarMut Fit")
                          )


```


# Selection coefficients across the phylogeny



```{r fig.height=7,fig.width=10}


deta.clust.all <- hclust(as.dist(1-cor(t(deta.matrix.for.heatmap),method="spearman",use="pairwise.complete.obs")),method="complete")
deta.dend.all <- as.dendrogram(deta.clust.all)


deta.phylo.hm <- Heatmap(deta.matrix.for.heatmap,
        col = col_fun_deta,   
        na_col = "black",
        cluster_rows = as.dendrogram(fungi.tree),
        name = "Selection Coefficient \u0394\u03B7", 
        #right_annotation = row_annot,
        left_annotation = clade_annot,
        show_row_names = F)

deta.clust.hm <- Heatmap(deta.matrix.for.heatmap,
        col = col_fun_deta,
        na_col = "black",
        cluster_rows = deta.dend.all,
        name = "Selection Coefficient \u0394\u03B7", 
        right_annotation = row_annot,
        left_annotation = clade_annot,
        split = 3,
        show_row_names = F)

deta.phylo.hm 
deta.clust.hm
```



```{r fig.width=10,fig.height=12}
deta.dend.clust <- cutree(deta.dend.all,k = 3)
table(deta.dend.clust)

ex.species <- c("saccharomyces_cerevisiae.max.cds","yHMPu5000034637_ogataea_populiabae_160519.max.cds","candida_sojae.max.cds")
# Note the ordering is reversed from the labeling in the ComplexHeatmap, so cluster 1 in cutree result is ComplexHeatmap cluster 3
phi.vs.emp.examples <- phi.vs.emp %>% 
  filter(Species %in% ex.species) %>%
  mutate(Species = case_when(
           Species == ex.species[1] ~ paste(Species = cleanSpeciesLabelsForPlots(Species),"(Cluster 3)"),
           Species == ex.species[2] ~ paste(Species = cleanSpeciesLabelsForPlots(Species),"(Cluster 2)"),
           Species == ex.species[3] ~ paste(Species = cleanSpeciesLabelsForPlots(Species),"(Cluster 1)")
  ))
  

emp.vs.phi.p <- ggplot(phi.vs.emp.examples,aes(x=Empirical,y=Phi)) +
  geom_point() +
  theme_cowplot() +
  scale_x_log10(labels = trans_format("log10", math_format(10^.x))) + 
  scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
  stat_cor(method="spearman",label.sep="\n",cor.coef.name = "rho") +
  scale_color_viridis_d() +
  labs(color="Hierarchical\ncluster") +
  ylab(expression("Scaled predicted expression ("*phi*")")) +
  xlab("Empirical expression (RNA TPM)") +
  theme(plot.title = element_blank()) +
  facet_wrap(~Species,nrow=3)


deta.vs.wi.examples <-  deta.vs.wi %>% 
  filter(Species %in% ex.species) %>%
  mutate(HClust = case_when(
    Species == ex.species[1] ~ "3",
    Species == ex.species[2] ~ "2",
    Species == ex.species[3] ~ "1"
    
  ),
  Species = cleanSpeciesLabelsForPlots(Species))



emp.vs.phi.p
```

## Compare correlations for clusters

```{r}
compare.hclust <- deta.dend.clust %>% 
  as.data.frame() %>% 
  rownames_to_column("Species") %>% 
  left_join(corr.df.ordered) %>% 
  dplyr::rename(HClust=".") %>% 
  mutate(HClust = ifelse(HClust == 1,1,2)) %>% 
  group_by(HClust) %>% 
  group_split()

t.test(compare.hclust[[1]]$Corr.Gene.Exp.vs.Phi,compare.hclust[[2]]$Corr.Gene.Exp.vs.Phi)
t.test(compare.hclust[[1]]$Corr.tAI.vs.Sel.Coef,compare.hclust[[2]]$Corr.tAI.vs.Sel.Coef)


```

```{r}
sphi.files <- list.files(path="../Final_runs",pattern="sphi.txt",full.names = T,recursive = T)
species.per.sphi.file <- lapply(str_split(sphi.files,pattern="/"),"[",4) %>% unlist()
names(sphi.files) <- species.per.sphi.file
sphi <- lapply(sphi.files,function(x)
  {
   data.frame(Sphi=as.numeric(readLines(x)))
  }) %>%
  bind_rows(.id="Species")


compare.hclust.sphi <- deta.dend.clust %>% 
  as.data.frame() %>% 
  rownames_to_column("Species") %>% 
  left_join(sphi) %>% 
  left_join(corr.df.ordered) %>%
  dplyr::rename(HClust=".") %>% 
  mutate(HClust = ifelse(HClust == 1,1,2)) %>% 
  group_by(HClust) %>% 
  group_split()


t.test(compare.hclust.sphi[[1]]$Sphi,compare.hclust.sphi[[2]]$Sphi)


```


```{r}
compare.hclust.sphi <- deta.dend.clust %>% 
  as.data.frame() %>% 
  rownames_to_column("Species") %>% 
  left_join(sphi) %>% 
  left_join(corr.df.ordered) %>%
  dplyr::rename(HClust=".") %>% 
  mutate(Sphi.low = ifelse(Sphi >= 0.5,1,2)) %>% 
  group_by(Sphi.low) %>% 
  group_split()

t.test(compare.hclust.sphi[[1]]$Corr.Gene.Exp.vs.Phi,compare.hclust.sphi[[2]]$Corr.Gene.Exp.vs.Phi)
t.test(compare.hclust.sphi[[1]]$Corr.tAI.vs.Sel.Coef,compare.hclust.sphi[[2]]$Corr.tAI.vs.Sel.Coef)


```



# Mutation bias $\Delta\mathit{M}$



## Mutation bias clustering with lower GC3% estimates

```{r fig.width=10,fig.height=10}
dm.clust.all <- hclust(as.dist(1-cor(t(dm.matrix.for.heatmap),method="spearman",use="pairwise.complete.obs")),method="complete")
dm.dend.all <- as.dendrogram(dm.clust.all)


dm.phylo.hm <- Heatmap(dm.matrix.for.heatmap,
        col = col_fun_dm,
        na_col = "black",
        cluster_rows = as.dendrogram(fungi.tree),
        name = "Mutation Bias \u0394M", 
        right_annotation = row_annot,
        left_annotation = clade_annot,
        show_row_names = F)

dm.clust.hm <- Heatmap(dm.matrix.for.heatmap,
        col = col_fun_dm,
        na_col = "black",
        cluster_rows = dm.dend.all,
        name = "Mutation Bias \u0394M", 
        right_annotation = row_annot,
        left_annotation = clade_annot,
        show_row_names = F)

cairo_pdf("../Figures/All_species/S06_heatmap_mutation_bias_phylogeny.pdf",width=10,height=7,family = "Arial Unicode MS")
draw(dm.phylo.hm)
dev.off()

cairo_pdf("../Figures/All_species/05_heatmap_mutation_bias_clust.pdf",width=10,height=7,family = "Arial Unicode MS")
draw(dm.clust.hm)
dev.off()


dm.phylo.hm
dm.clust.hm

```

```{r fig.width=18,fig.height=18}
layout <- "
ABBBBC
ABBBBC
ADDDDC
ADDDD#
"


p <- (plot_spacer() + grid.grabExpr(draw(deta.clust.hm),wrap = T,wrap.grobs = T)) + emp.vs.phi.p + (grid.grabExpr(draw(dm.clust.hm),wrap = T,wrap.grobs = T))  +
  plot_annotation(tag_levels = "A") + 
  plot_layout(widths = c(0,2,1),design = layout) 
p

savePlot("../Figures/All_species/heatmap_selection_coefficients_mutation_bias_constmut_lower_gc3_clust.pdf",
       plot = p,
       height=16,
       width=18)


```

## Mutation bias clustering with higher GC3% estimates

```{r fig.width=10,fig.height=10}
dm.matrix.for.heatmap <- dm.matrix
dm.matrix.for.heatmap <- dm.matrix.for.heatmap %>%
  dplyr::slice(match(fungi.tree.higher.gc3$tip.label,Species)) %>%
  mutate(Species = str_remove(Species,"_Higher_GC3")) %>%
  column_to_rownames("Species") %>% 
  as.matrix()

colnames(dm.matrix.for.heatmap) <- aa.codon


dm.clust.all <- hclust(as.dist(1-cor(t(dm.matrix.for.heatmap),method="spearman",use="pairwise.complete.obs")),method="complete")
dm.dend.all <- as.dendrogram(dm.clust.all)

dm.phylo.hm <- Heatmap(dm.matrix.for.heatmap,
        col = col_fun_dm,
        na_col = "black",
        cluster_rows = as.dendrogram(fungi.tree),
        name = "Mutation Bias", 
        right_annotation = row_annot,
        left_annotation = clade_annot,
        show_row_names = F)

dm.clust.hm <- Heatmap(dm.matrix.for.heatmap,
        col = col_fun_dm,
        na_col = "black",
        cluster_rows = dm.dend.all,
        name = "Mutation Bias", 
        right_annotation = row_annot,
        left_annotation = clade_annot,
        show_row_names = F)

cairo_pdf("../Figures/All_species/heatmap_mutation_bias_higher_gc3_phylogeny.pdf",width=10,height=7,family="Arial Unicode MS")
draw(dm.phylo.hm)
dev.off()

cairo_pdf("../Figures/All_species/heatmap_mutation_bias_higher_gc3_clust.pdf",width=10,height=7,family="Arial Unicode MS")
draw(dm.clust.hm)
dev.off()


dm.phylo.hm
dm.clust.hm

```


```{r}
deta.dend.clust <- cutree(deta.dend.all,k = 3)
table(deta.dend.clust)
questionable <- names(deta.dend.clust)[which(deta.dend.clust == 2 | deta.dend.clust == 3)]
write(questionable,"../Post_analysis/poor_roc_fits_based_on_hierarchical_clustering.txt",ncolumns = 1)


fungi.tree.deta.filt <- drop.tip(fungi.tree, questionable)
fungi.tree.dm.filt <- drop.tip(fungi.tree.deta.filt,questionable)
fungi.tree.dm.1.filt <- drop.tip(fungi.tree.deta.filt,questionable)
fungi.tree.dm.2.filt <- drop.tip(fungi.tree.deta.filt,questionable)

for (i in setdiff(improved.fits.species,questionable))
{
  fungi.tree.dm.filt <- bind.tip(fungi.tree.dm.filt,tip.label=paste0(i,"_Higher_GC3"),
                              where = which(fungi.tree.dm.filt$tip.label == i))
  fungi.tree.dm.filt$tip.label[which(fungi.tree.dm.filt$tip.label == i)] <- paste0(i,"_Lower_GC3")
  fungi.tree.dm.1.filt$tip.label[which(fungi.tree.dm.1.filt$tip.label == i)] <- paste0(i,"_Lower_GC3")
  fungi.tree.dm.2.filt$tip.label[which(fungi.tree.dm.2.filt$tip.label == i)] <- paste0(i,"_Higher_GC3")
}

```



# Compare hierarchical clustering of selection coefficients and mutation biases to phylogeny

```{r,fig.width=20,fig.height=20}
phy.tree.dend <- as.dendrogram(fungi.tree.deta.filt)

deta.matrix.for.heatmap <- deta.matrix
deta.matrix.for.heatmap <- deta.matrix.for.heatmap %>%
  dplyr::slice(match(fungi.tree.deta.filt$tip.label,Species)) %>%
  column_to_rownames("Species") %>% 
  as.matrix()

sd.deta.matrix.for.heatmap <- sd.deta.matrix
sd.deta.matrix.for.heatmap <- sd.deta.matrix.for.heatmap %>%
  dplyr::slice(match(fungi.tree.deta.filt$tip.label,Species)) %>%
  column_to_rownames("Species") %>% 
  as.matrix()

deta.clust.filt <- hclust(as.dist(1-cor(t(deta.matrix.for.heatmap),method="spearman",use="pairwise.complete.obs")),method="complete")
deta.dend.filt <- as.dendrogram(deta.clust.filt)


dm.matrix.low.gc.for.heatmap <- dm.matrix
dm.matrix.low.gc.for.heatmap <- dm.matrix.low.gc.for.heatmap %>%
  dplyr::slice(match(fungi.tree.dm.1.filt$tip.label,Species)) %>%
  mutate(Species = str_remove(Species,"_Lower_GC3")) %>%
  column_to_rownames("Species") %>% 
  as.matrix()

sd.dm.matrix.low.gc.for.heatmap <- sd.dm.matrix
sd.dm.matrix.low.gc.for.heatmap <- sd.dm.matrix.low.gc.for.heatmap %>%
  dplyr::slice(match(fungi.tree.dm.1.filt$tip.label,Species)) %>%
  mutate(Species = str_remove(Species,"_Lower_GC3")) %>%
  column_to_rownames("Species") %>% 
  as.matrix()

dm.clust.filt <- hclust(as.dist(1-cor(t(dm.matrix.low.gc.for.heatmap),method="spearman",use="pairwise.complete.obs")),method="complete")
dm.dend.filt <- as.dendrogram(dm.clust.filt)


deta.dend.list <- dendlist(deta.dend.filt,phy.tree.dend)
corr <- cor.dendlist(deta.dend.list,method="cophenetic")
print(paste("Cophenetic correlation between Selection Coefficient Clustering and Phylogenetic Tree:",round(corr[1,2], 2)))


dm.dend.list <- dendlist(dm.dend.filt,phy.tree.dend)
corr <- cor.dendlist(dm.dend.list,method="cophenetic")
print(paste("Cophenetic correlation between Mutation Bias (Lower GC3%) Clustering and Phylogenetic Tree:",round(corr[1,2], 2)))

dm.matrix.high.gc.for.heatmap <- dm.matrix
dm.matrix.high.gc.for.heatmap <- dm.matrix.high.gc.for.heatmap %>%
  dplyr::slice(match(fungi.tree.dm.2.filt$tip.label,Species)) %>%
  mutate(Species = str_remove(Species,"_Higher_GC3")) %>%
  column_to_rownames("Species") %>% 
  as.matrix()

sd.dm.matrix.high.gc.for.heatmap <- sd.dm.matrix
sd.dm.matrix.high.gc.for.heatmap <- sd.dm.matrix.high.gc.for.heatmap %>%
  dplyr::slice(match(fungi.tree.dm.2.filt$tip.label,Species)) %>%
  mutate(Species = str_remove(Species,"_Higher_GC3")) %>%
  column_to_rownames("Species") %>% 
  as.matrix()

dm.clust.filt <- hclust(as.dist(1-cor(t(dm.matrix.high.gc.for.heatmap),method="spearman",use="pairwise.complete.obs")),method="complete")
dm.dend.filt <- as.dendrogram(dm.clust.filt)

dm.dend.list <- dendlist(dm.dend.filt,phy.tree.dend)
corr <- cor.dendlist(dm.dend.list,method="cophenetic")
print(paste("Cophenetic correlation between Mutation Bias (Higher GC3%) Clustering and Phylogenetic Tree:",round(corr[1,2], 2)))


```


Here, we calculate Blomberg's $K$ using the `phylosig` function from `phytools` for individual codons. This analysis is superseded by the multivariate Blomberg's $K$ analysis below.

```{r}

codon.vec <- colnames(deta.matrix.for.heatmap)

deta.k.dist <- codon.vec %>% 
  purrr::map(~phylosig(tree = fungi.tree.deta.filt,
                       x=deta.matrix.for.heatmap[,.x],
                       se = sd.deta.matrix.for.heatmap[,.x],
                       method="K")
  
) %>% purrr::map(~.x$K) %>%
  unlist()


dm.low.gc.k.dist <- codon.vec %>% 
  purrr::map(~phylosig(tree = fungi.tree.deta.filt,
                       x= dm.matrix.low.gc.for.heatmap[,.x],
                       se = sd.dm.matrix.low.gc.for.heatmap[,.x],
                       method="K")
  
) %>% purrr::map(~.x$K) %>%
  unlist()


dm.high.gc.k.dist <- codon.vec %>% 
  purrr::map(~phylosig(tree = fungi.tree.deta.filt,
                       x= dm.matrix.high.gc.for.heatmap[,.x],
                       se = sd.dm.matrix.high.gc.for.heatmap[,.x],
                       method="K")
  
) %>% purrr::map(~.x$K) %>%
  unlist()


k.df <- data.frame(Codon = codon.vec, 
                   Selection = deta.k.dist, 
                   Mutation.Lower.GC3 = dm.low.gc.k.dist, 
                   Mutation.Higher.GC3 = dm.high.gc.k.dist)

k.comp <- ggplot(k.df,aes(x=Selection,y=Mutation.Lower.GC3,label=Codon)) +
  geom_point() +
  geom_text_repel() +
  geom_abline(intercept=0,slope=1,linetype="dashed") +
  stat_smooth(formula = y ~ x,method="lm") +
  stat_cor(method="spearman",label.sep="\n") +
  xlab("Blomberg's K\nSelection Coefficient") +
  ylab("Blomberg's K\nMutation Bias (Lower GC3% Set)") +
  theme_cowplot() +
  theme(aspect.ratio = 1)


k.comp


```


```{r}
deta.matrix.for.geomorph <- deta.matrix
deta.matrix.for.geomorph <- deta.matrix.for.geomorph %>%
  dplyr::slice(match(fungi.tree.deta.filt$tip.label,Species)) %>%
  column_to_rownames("Species") %>% 
  dplyr::select(-CTG) %>%
  as.matrix()


deta.k.multi <- geomorph::physignal(deta.matrix.for.geomorph,phy=fungi.tree.deta.filt)

dm.matrix.for.geomorph <- dm.matrix
dm.matrix.for.geomorph <- dm.matrix.for.geomorph %>%
  dplyr::slice(match(fungi.tree.dm.1.filt$tip.label,Species)) %>%
  mutate(Species = str_remove(Species,"_Lower_GC3")) %>%
  column_to_rownames("Species") %>% 
  dplyr::select(-CTG) %>%
  as.matrix()

dm.lower.gc3.k.multi <- geomorph::physignal(dm.matrix.for.geomorph,phy=fungi.tree.deta.filt)


dm.matrix.for.geomorph <- dm.matrix
dm.matrix.for.geomorph  <- dm.matrix.for.geomorph  %>%
  dplyr::slice(match(fungi.tree.dm.2.filt$tip.label,Species)) %>%
  mutate(Species = str_remove(Species,"_Higher_GC3")) %>%
  column_to_rownames("Species") %>% 
  dplyr::select(-CTG) %>%
  as.matrix()

dm.higher.gc3.k.multi <- geomorph::physignal(dm.matrix.for.geomorph,phy=fungi.tree.deta.filt)

deta.k.multi
dm.lower.gc3.k.multi
dm.higher.gc3.k.multi


```




```{r}
codon.vec <- colnames(deta.matrix.for.heatmap)

names(codon.vec) <- codon.vec

fitContinuousWrapper <- function(phy,dat,...)
{
  dat.no.na <- dat[which(!is.na(dat))]
  cleaned <- cleanData(phy,dat.no.na)
  phy.clean <- cleaned$phy
  dat.clean <- cleaned$dat
  fitContinuous(phy.clean,dat.clean,...)
}


deta.evo.rate <- codon.vec %>% 
  purrr::map(~fitContinuousWrapper(phy = fungi.tree.deta.filt,
                       dat = deta.matrix.for.heatmap[,.x],
                       SE = sd.deta.matrix.for.heatmap[,.x],
                       model = "OU",
                       control=list(hessian=T))
) %>% purrr::map(~data.frame(Alpha=.x$opt$alpha,
                             Alpha.CI.LB = .x$opt$CI[1,"alpha"],
                             Alpha.CI.UB = .x$opt$CI[2,"alpha"],
                             SigSq=.x$opt$sigsq,
                             SigSq.CI.LB = .x$opt$CI[1,"sigsq"],
                             SigSq.CI.UB = .x$opt$CI[2,"sigsq"])) %>%
  bind_rows(.id="Codon")


dm.low.gc.evo.rate  <- codon.vec %>% 
  purrr::map(~fitContinuousWrapper(phy = fungi.tree.deta.filt,
                       dat = dm.matrix.low.gc.for.heatmap[,.x],
                       SE = sd.dm.matrix.low.gc.for.heatmap[,.x],
                       model = "OU",
                       control=list(hessian=T))
)  %>% purrr::map(~data.frame(Alpha=.x$opt$alpha,
                             Alpha.CI.LB = .x$opt$CI[1,"alpha"],
                             Alpha.CI.UB = .x$opt$CI[2,"alpha"],
                             SigSq=.x$opt$sigsq,
                             SigSq.CI.LB = .x$opt$CI[1,"sigsq"],
                             SigSq.CI.UB = .x$opt$CI[2,"sigsq"])) %>%
  bind_rows(.id="Codon")

dm.high.gc.evo.rate <- codon.vec %>% 
  purrr::map(~fitContinuousWrapper(phy = fungi.tree.deta.filt,
                       dat = dm.matrix.high.gc.for.heatmap[,.x],
                       SE = sd.dm.matrix.high.gc.for.heatmap[,.x],
                       model = "OU",
                       control=list(hessian=T))
) %>% purrr::map(~data.frame(Alpha=.x$opt$alpha,
                             Alpha.CI.LB = .x$opt$CI[1,"alpha"],
                             Alpha.CI.UB = .x$opt$CI[2,"alpha"],
                             SigSq=.x$opt$sigsq,
                             SigSq.CI.LB = .x$opt$CI[1,"sigsq"],
                             SigSq.CI.UB = .x$opt$CI[2,"sigsq"])) %>%
  bind_rows(.id="Codon")



ou.list <- list(deta.evo.rate,
              dm.low.gc.evo.rate,
              dm.high.gc.evo.rate)
parameter.names <- list("Selection","Mutation.Lower.GC3","Mutation.Higher.GC3")

ou.df <- purrr::map2(ou.list,parameter.names,function(x,y) 
  {
    x %>% rename_with(.fn = ~paste(.,y,sep="."), .cols = -Codon)
  }
)


ou.df <- ou.df %>%
  purrr::reduce(left_join,by="Codon")
```


```{r}
alpha.deta.vs.dm.low.gc <- ggplot(ou.df,aes(x=Alpha.Selection,y=Alpha.Mutation.Lower.GC3)) +
  geom_point() +
  geom_errorbar(aes(ymin=Alpha.CI.LB.Mutation.Lower.GC3,ymax=Alpha.CI.UB.Mutation.Lower.GC3),alpha=0.5) +
  geom_errorbarh(aes(xmin=Alpha.CI.LB.Selection,xmax=Alpha.CI.UB.Selection),alpha=0.5) +
  geom_abline(intercept=0,slope=1) +
  xlab("Stabilizing Selection Strength\nSelection Coefficients") +
  ylab("Stabilizing Selection Strength\nMutation Bias (Lower GC3% Set)") +
  theme_cowplot() +
  theme(aspect.ratio=1)
alpha.deta.vs.dm.low.gc

alpha.deta.vs.dm.high.gc <- ggplot(ou.df,aes(x=Alpha.Selection,y=Alpha.Mutation.Higher.GC3)) +
  geom_point() +
  geom_errorbar(aes(ymin=Alpha.CI.LB.Mutation.Higher.GC3,ymax=Alpha.CI.UB.Mutation.Higher.GC3),alpha=0.5) +
  geom_errorbarh(aes(xmin=Alpha.CI.LB.Selection,xmax=Alpha.CI.UB.Selection),alpha=0.5) +
  geom_abline(intercept=0,slope=1) +
  xlab("Stabilizing Selection Strength\nSelection Coefficient") +
  ylab("Stabilizing Selection Strength\nMutation Bias (Higher GC3% Set)") +
  theme_cowplot() +
  theme(aspect.ratio=1)
alpha.deta.vs.dm.high.gc
  

alpha.comp <- (alpha.deta.vs.dm.low.gc | alpha.deta.vs.dm.high.gc) + plot_annotation(tag_levels = "A")
alpha.comp

ggsave2("../Figures/All_species/ou_alpha_comparisons.pdf",width=8,height=5)
```



