---
title: "Evolution of mutation bias across species"
author: "Alex Cope"
date: "1/26/2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ape)
library(AnaCoDa)
library(phytools)
library(tidyverse)
library(cowplot)
library(reshape2)
library(ggpubr)
library(broom)
library(ggtree)
library(patchwork)
library(ComplexHeatmap)
library(tidytext)
library(phylolm)
source("helperFunctions.R")
```

# Set up color schemes

```{r}
scale_colour_brewer_d <- function(..., palette = "Set3") {
  scale_colour_brewer(..., palette = palette )
}

scale_fill_brewer_d <- function(..., palette = "Set1") {
  scale_fill_brewer(..., palette = palette)
  
}

scale_colour_brewer_c <- function(..., palette = "BuGn") {
  scale_colour_distiller(..., palette = palette )
}

scale_fill_brewer_c <- function(..., palette = "BuGn") {
  scale_fill_distiller(..., palette = palette)
}

options(
  ggplot2.discrete.colour = scale_colour_brewer_d,
  ggplot2.discrete.fill = scale_fill_brewer_d,
  ggplot2.continuous.colour = scale_colour_brewer_c,
  ggplot2.continuous.fill = scale_fill_brewer_c
)
```

# Read in the relevant data

## Phylogeny

```{r}
all.genomes <- readLines("../Data/all_fungi.txt")
fungi.tree <- read.tree("../Data/tree_with_cds_labels.nwk")
fungi.tree <- cleanTree(fungi.tree,all.genomes)
clades <- read_tsv("../Data/clades.txt")
clades <- clades %>% 
  dplyr::slice(match(fungi.tree$tip.label,file_name_id)) %>%
  dplyr::rename(Species=file_name_id,Clade=Major.clade)



improved.fits.species <- readLines("../Data/improved_species_by_gene_expression_25_percent_2023_05_03.txt")
poor.fits <- readLines("../Post_analysis/poor_roc_fits_based_on_hierarchical_clustering.txt")

fungi.tree.filt <- drop.tip(fungi.tree, poor.fits)
fungi.tree.filt.lower.gc3 <- fungi.tree.filt
fungi.tree.filt.higher.gc3 <- fungi.tree.filt

for (i in setdiff(improved.fits.species,poor.fits))
{
  fungi.tree.filt.lower.gc3$tip.label[which(fungi.tree.filt.lower.gc3$tip.label == i)] <- paste0(i,"_Lower_GC3")
  fungi.tree.filt.higher.gc3$tip.label[which(fungi.tree.filt.higher.gc3$tip.label == i)] <- paste0(i,"_Higher_GC3")
}


```

## Mutation bias and selection coefficients estimates across 327 yeasts

For comparing to GC%, we will use codon-specific parameter where each estimate is scaled such that G is relative to A and C is relative to T.

```{r}
deta.matrix.k1 <- read_tsv("../Post_analysis/selection_coefficients_k1.tsv",col_types = cols())
deta.matrix.k2 <- read_tsv("../Post_analysis/selection_coefficients_k2.tsv",col_types = cols())

deta.matrix.k1.rescaled <- read_tsv("../Post_analysis/selection_coefficients_k1_GC_rescaled.tsv",col_types = cols())
deta.matrix.k2.rescaled <- read_tsv("../Post_analysis/selection_coefficients_k2_GC_rescaled.tsv",col_types = cols())


dm.matrix.k1 <- read_tsv("../Post_analysis/mutation_bias_k1.tsv",col_types = cols())
dm.matrix.k2.higher.gc3 <- read_tsv("../Post_analysis/mutation_bias_k2_higher_GC3.tsv",col_types = cols())
dm.matrix.k2.lower.gc3 <- read_tsv("../Post_analysis/mutation_bias_k2_lower_GC3.tsv",col_types = cols())


dm.matrix.k1.rescaled <- read_tsv("../Post_analysis/mutation_bias_k1_GC_rescaled.tsv",col_types = cols())
dm.matrix.k2.higher.gc3.rescaled <- read_tsv("../Post_analysis/mutation_bias_k2_higher_GC3_GC_rescaled.tsv",col_types = cols())
dm.matrix.k2.lower.gc3.rescaled <- read_tsv("../Post_analysis/mutation_bias_k2_lower_GC3_GC_rescaled.tsv",col_types = cols())


```


## Create final parameter matrices dependent on improved species fits 

```{r}
deta.matrix <- deta.matrix.k1
deta.matrix <- deta.matrix %>% 
  column_to_rownames("Species")
deta.matrix <- as.data.frame(t(deta.matrix))


dm.matrix <- dm.matrix.k1
dm.matrix <- dm.matrix %>% 
  column_to_rownames("Species")
dm.matrix <- as.data.frame(t(dm.matrix))


deta.matrix.gc <- deta.matrix.k1.rescaled
deta.matrix.gc <- deta.matrix.gc %>% 
  column_to_rownames("Species")
deta.matrix.gc <- as.data.frame(t(deta.matrix.gc))

dm.matrix.gc <- dm.matrix.k1.rescaled
dm.matrix.gc <- dm.matrix.gc %>% 
  column_to_rownames("Species")
dm.matrix.gc <- as.data.frame(t(dm.matrix.gc))



for (i in improved.fits.species)
{
  deta.matrix <- deta.matrix %>%
    mutate(!!as.name(i) := deta.matrix.k2 %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    )
  
  deta.matrix.gc <- deta.matrix.gc %>% 
    mutate(!!as.name(i) := deta.matrix.k2.rescaled %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    )
  
  dm.matrix <- dm.matrix %>% 
    mutate(!!as.name(i) := dm.matrix.k2.higher.gc3%>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist(),
           !!as.name(paste0(i,"_Lower_GC3")) := dm.matrix.k2.lower.gc3 %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    ) %>%
      dplyr::rename(!!as.name(paste0(i,"_Higher_GC3")) := !!as.name(i))
  
  
  dm.matrix.gc <- dm.matrix.gc %>% 
    mutate(!!as.name(i) := dm.matrix.k2.higher.gc3.rescaled%>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist(),
           !!as.name(paste0(i,"_Lower_GC3")) := dm.matrix.k2.lower.gc3.rescaled %>% 
            filter(Species == i) %>% 
            dplyr::select(-Species) %>%
            unlist()
    ) %>%
      dplyr::rename(!!as.name(paste0(i,"_Higher_GC3")) := !!as.name(i))
}

param.list <- list(deta.matrix,deta.matrix.gc,dm.matrix.gc,dm.matrix) %>%
  purrr::map(~.x %>%
              t() %>% 
              as.data.frame() %>%
              rownames_to_column("Species") %>%
              as_tibble())

deta.matrix <- param.list[[1]]
deta.matrix.gc <- param.list[[2]]
dm.matrix.gc <- param.list[[3]]
dm.matrix <- param.list[[4]]


```

## ROC-SEMPPR $\phi$

```{r}
sphi <- lapply(all.genomes,function(species){
  if (species %in% improved.fits.species)
  {
    data.frame(Species = species,Sphi = as.numeric(readLines(file.path("..","Final_runs","Results_k_2_selectionShared",species,"run_2","Parameter_est","sphi.txt"))))
  } else {
    data.frame(Species = species,Sphi=as.numeric(readLines(file.path("..","Final_runs","Results_k_1",species,"run_2","Parameter_est","sphi.txt"))))
  }
}) %>% bind_rows()


```

```{r}
phi <- read_csv("../Data/phi_shen_etal_ortholog_matrix_2023_05_01.csv")

phi <- phi %>% 
  column_to_rownames("Orthogroup") %>%
  rename_with(~cleanColumnNames(.x,fungi.tree$tip.label))
phi <- phi[,fungi.tree$tip.label]

phi.transform <- phi %>% 
  mutate(across(where(is.numeric),log))

phi.transform <- as.data.frame(normalize.quantiles(as.matrix(phi.transform),keep.names = T))

phi.filt <- phi.transform[which(rowSums(is.na(phi.transform)) < 120),]
phi <- t(phi.filt)

```


```{r}
phi.long <- phi.transform %>%
  pivot_longer(everything(),names_to="Species",values_to = "Phi")
ggplot(phi.long,aes(x=Species,y=Phi)) +
  geom_boxplot()


```

```{r echo = F, fig.width=9,fig.height=9}
gc.cont <- read_tsv("../Data/gc_content_by_genome.tsv") %>%
           dplyr::rename(Species=Genome) %>%
           dplyr::slice(match(fungi.tree$tip.label,Species)) %>%
           left_join(clades,by="Species")



```

```{r}
deta.vs.wi <- read_tsv("../Post_analysis/roc_sel_coef_vs_tAI_for_all_327_yeasts.tsv") 

deta.gc.wi <- deta.vs.wi %>%
  left_join(gc.cont,by="Species")



```


# Compare selection coefficients and mutation bias to GC% across species

```{r fig.width=12, warning = F}
deta.tree.to.use <- fungi.tree.filt
dm.tree.to.use <- fungi.tree.filt.lower.gc3

deta.matrix.t <- deta.matrix.gc %>%
  filter(Species %in% deta.tree.to.use$tip.label)
deta.matrix.t <- as.data.frame(deta.matrix.t)


dm.matrix.t <- dm.matrix.gc %>%
  filter(Species %in% dm.tree.to.use$tip.label)
dm.matrix.t <- as.data.frame(dm.matrix.t)

# deta.matrix.t <- deta.matrix %>%
#   filter(Species %in% deta.tree.to.use$tip.label)
# deta.matrix.t <- as.data.frame(deta.matrix.t)
#   
# 
# dm.matrix.t <- dm.matrix %>%
#   filter(Species %in% dm.tree.to.use$tip.label)
# dm.matrix.t <- as.data.frame(dm.matrix.t)
#   

deta.matrix.long <- deta.matrix.t %>% 
                 dplyr::slice(match(deta.tree.to.use$tip.label,Species)) %>%
                 pivot_longer(-Species,names_to="Codon",values_to="Selection") 


dm.matrix.long <- dm.matrix.t %>% 
                 mutate(Species = str_remove(Species,"_Lower_GC3$")) %>%
                 dplyr::slice(match(deta.tree.to.use$tip.label,Species)) %>%
                 pivot_longer(-Species,names_to="Codon",values_to="Mutation") 

dm.gc <- dm.matrix.long %>%
                 inner_join(deta.matrix.long,by=c("Species","Codon")) %>%
                 inner_join(gc.cont,by="Species")# %>%
                 #mutate(Codon=factor(Codon,levels=AnaCoDa::codons()))
                

cor.gc.pic <- dm.gc %>% 
  group_by(Codon,Species) %>% 
  nest(Species,Mutation,Selection,GC,Clade) %>% 
  mutate(Mutation = purrr::map(data,function(x)
    {
      calculateAcrossSpeciesCorrelations(x,tree=deta.tree.to.use,x.var="GC",y.var="Mutation",method="spearman")
    }
    )) %>% 
  dplyr::select(-data) %>% 
  unnest()


cor.deta.pic <- dm.gc %>%
  group_by(Codon,Species) %>%
  nest(Species,Mutation,Selection,GC,Clade) %>%
  mutate(Mutation = purrr::map(data,function(x)
    {
      calculateAcrossSpeciesCorrelations(x,tree=deta.tree.to.use,"Mutation","Selection",method="spearman")
    }
    )) %>%
  dplyr::select(-data) %>%
  unnest()


cor.gc.pic <- cor.gc.pic %>% dplyr::select(Codon,contains("estimate"),
                                            contains("p.value")) %>%
          ungroup() %>%
          mutate(adj.p.value = p.adjust(p.value,method="BH")
                 ) %>%
          mutate(Significant = ifelse(adj.p.value < 0.05,"*","")) %>% 
           rowwise() %>%
                 mutate(AA = AnaCoDa::codonToAA(as.character(Codon)),
                        Num.Codons = as.character(length(AnaCoDa::AAToCodon(AA)))) %>%
                 ungroup() %>%
                 mutate(Codon.AA = paste(AA,Codon,sep="-"))


cor.deta.pic <- cor.deta.pic %>% dplyr::select(Codon,contains("estimate"),
                                            contains("p.value")) %>%
          ungroup() %>%
          mutate(adj.p.value = p.adjust(p.value,method="BH")
                 ) %>%
          mutate(Significant = ifelse(adj.p.value < 0.05,"*","")) %>%
           rowwise() %>%
                 mutate(AA = AnaCoDa::codonToAA(as.character(Codon)),
                        Num.Codons = as.character(length(AnaCoDa::AAToCodon(AA)))) %>%
                 ungroup() %>%
                 mutate(Codon.AA = paste(AA,Codon,sep="-"))




```


```{r fig.width=12,fig.height=5}

target.codon <- c("CAG")
dm.gc <- dm.gc %>% 
  mutate(Codon=as.character(Codon))

dm.gc.pic <- dm.gc %>% 
  split(f = ~as.factor(Codon)) %>%
  purrr::map(function(x)
    {
      tmp <- x %>% column_to_rownames("Species") %>% 
        dplyr::select(GC,Mutation) %>% 
        dplyr::filter(!is.na(GC) & !is.na(Mutation)) %>%
        as.matrix()
      cleaned <- cleanData(deta.tree.to.use,tmp)
      x <- cleaned$data
      tmp.tree <- cleaned$phy
      gc.values <- x[,"GC"]
      mutation.values <- x[,"Mutation"]
      gc.pic <- calculatePIC(gc.values[tmp.tree$tip.label],tree = tmp.tree)
      mut.pic <- calculatePIC(mutation.values[tmp.tree$tip.label],tree=tmp.tree)
      df <- data.frame(GC.pic = gc.pic[,1],Mutation.pic=mut.pic[,1])
      df
    }
    ) %>% bind_rows(.id="Codon") 

dm.gc.pic.target <- dm.gc.pic %>% 
  filter(Codon %in% target.codon)

gc.mut <- ggplot(dm.gc.pic.target ,aes(x=GC.pic,y=Mutation.pic)) +
          geom_point() +
          xlab("GC% (PIC)") +
          ylab("Mutation bias (PIC)") +
          ggtitle("Q-CAG") +
          stat_cor(method="spearman",label.sep="\n") +
          theme_cowplot() 
gc.mut


bar.p.gc <- ggplot(cor.gc.pic,aes(x=reorder(Codon.AA, estimate),y=estimate,shape=Significant,fill=Num.Codons)) + 
  geom_bar(stat="identity",position = position_dodge(),color="black",fill="white") + 
  geom_point(aes(y = ifelse(estimate > 0, estimate + 0.05, estimate - 0.05)),
             position=position_dodge(0.9),
             show.legend=FALSE) +
  scale_shape_manual(values = c(8)) +
  xlab("Codon") +
  ylab("Spearman Correlation") +
  ylim(c(-0.8,0.1)) +
  geom_hline(yintercept = 0) + 
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
bar.p.gc
```


```{r}
layout <- "
AABBBB
"
mut.plot <- gc.mut + bar.p.gc + plot_annotation(tag_levels = "A") + plot_layout(design = layout)
mut.plot


savePlot("../Figures/All_species/mutation_bias_vs_gc_content.pdf",mut.plot,width=10,height=4,family="sans")

```
A similar result is obtained if using PGLS.

```{r}
dm.gc.split.by.codon <- dm.gc %>% 
  filter(Codon %in% colnames(deta.matrix)) %>%
  left_join(sphi) %>%
  group_by(Species) %>%
  ungroup() %>%
  split(f=~Codon)


pgls.mut.gc.pic <- lapply(dm.gc.split.by.codon,function(x) {
      tryCatch({
        x <- x %>%
        column_to_rownames("Species")
        x <- x[deta.tree.to.use$tip.label,]
        plm.lambda <- phylolm(GC ~ Mutation + Sphi,
                              data= x, 
                              phy = deta.tree.to.use,
                              model = "lambda")
        model.sum <- summary(plm.lambda)
        model.coef <- model.sum$coefficients
        df <- model.coef %>% 
          as.data.frame() %>% 
          rownames_to_column("Coefficient") %>% 
          dplyr::select(-t.value) %>% 
          mutate(Evo.Model = plm.lambda$model)
        return(df)
      },
      error=function(e){return(NA)},
      warnings = function(w){})
    }
    )

pgls.mut.gc.pic <- pgls.mut.gc.pic %>%  
  bind_rows(.id="Codon")



```

```{r}
all.slope.pgls <- pgls.mut.gc.pic %>% 
  ungroup() %>%
  mutate(adj_p.value = p.adjust(p.value,method="BH"),
        Codon=factor(Codon,levels=AnaCoDa::codons()),
        Significant=ifelse(adj_p.value < 0.05,"*","")) %>%
  filter(Coefficient == "Mutation")

slope.bar.p.all <- ggplot(all.slope.pgls,aes(x=reorder(Codon, Estimate),y=Estimate,shape=Significant)) + 
  geom_bar(stat="identity",position = position_dodge(),color="black",fill="white") + 
  geom_point(aes(y = ifelse(Estimate > 0, Estimate + StdErr + 0.05, Estimate - StdErr - 0.05)),
             position=position_dodge(0.9),
             show.legend=FALSE) +
  theme_cowplot() +
  geom_errorbar(aes(ymin = Estimate - StdErr,ymax=Estimate + StdErr)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ylab("PGLS Slope") +
  labs(fill = "Selection coefficient\n-waiting time correlation") +
  #ggtitle("Across-species correlations between elongation waiting times\nand selection coefficients \u0394\u03B7") +
  xlab("Codon") +
  scale_shape_manual(values = c(NA, 8))
slope.bar.p.all



```




# Correlation between mutation bias and shifts in mismatch repair gene expression

```{r}
possible.double.counts <- c("PCN1","RHP41","RHP42","CDC1","CDC27","CDC6","SSB1","SSB2","ADL1","CDC17","POLD1","POLD2","POLD3","POLD4","RPA3","PMS2","LIG1")

mmr <- read_tsv("../Data/mmr_matrix.tsv") 
all.genes <- colnames(mmr[2:ncol(mmr)])
keep.genes <- all.genes[which(!all.genes %in% possible.double.counts)] 
species.original <- deta.tree.to.use$tip.label
species <- cleanSpeciesLabels(species.original)

mmr<-mmr[,c("Species",keep.genes)]

num.genes <- ncol(mmr) - 1

mmr.sacch <- mmr %>% 
  filter(Species %in% species) %>%
  mutate(Missing = num.genes - rowSums(across(where(is.numeric))))

fungi.tree.cleaned <- deta.tree.to.use
fungi.tree.cleaned$tip.label <- species
missing.species <- fungi.tree.cleaned$tip.label[which(!fungi.tree.cleaned$tip.label %in% mmr.sacch$Species)]
fungi.tree.cleaned <- drop.tip(fungi.tree.cleaned,missing.species)

dup.species <- fungi.tree.cleaned$tip.label[which(duplicated(fungi.tree.cleaned$tip.label))]
fungi.tree.cleaned <- drop.tip(fungi.tree.cleaned,dup.species)

mmr.sacch <- mmr.sacch %>%
  distinct(Species,.keep_all = T)

gc.cont.clean <- gc.cont %>%
  mutate(Species = cleanSpeciesLabels(Species))

mmr.sacch <- mmr.sacch %>%
  left_join(gc.cont.clean,by="Species") %>%
  dplyr::slice(match(fungi.tree.cleaned$tip.label,Species))



gc.comp.all <- ggplot(mmr.sacch,aes(x=as.factor(Missing),y=GC)) +
  geom_boxplot(color="black") +
  geom_jitter(color="black", size=0.4, alpha=0.9) +
  xlab("Number of missing MMR genes") +
  ylab("GC%") +
  #ggtitle("Impact of missing MMRs on GC%") +
  theme_cowplot()
gc.comp.all 

phy.reg <- phylolm(GC ~ Missing,data=mmr.sacch %>% column_to_rownames("Species"),phy=fungi.tree.cleaned,model="lambda")

summary(phy.reg)

```


```{r}

#yeast_orthologs <- read_delim("../Data/mmr_gene_ids.csv",delim=" ") %>% 
                    # filter(reason == "MATCH" | reason == "OTHER") %>%
                    # dplyr::select(input,secondaryIdentifier) %>%
                    # distinct(secondaryIdentifier,.keep_all = T)
yeast_orthologs <- read_tsv("../Data/mmr_gene_ids_updated.tsv") %>%
  dplyr::rename(input=Short_Name)


orthogroups <- read_tsv("../Data/shen_etal_ortholog_matrix_2021_12_17.tsv")

orthogroups <- orthogroups %>% 
       column_to_rownames("Orthogroup") %>%
       rename_with(~cleanColumnNames(.x,fungi.tree$tip.label))

orthogroups<- orthogroups %>% 
  rownames_to_column("Orthogroup")

mmr.ortho <- yeast_orthologs %>% 
  #left_join(orthogroups,by=c("secondaryIdentifier" = "saccharomyces_cerevisiae.max.cds")) %>%
  left_join(orthogroups,by=c("Systematic_Name" = "saccharomyces_cerevisiae.max.cds")) %>%
  #dplyr::select(input,secondaryIdentifier,Orthogroup) 
  dplyr::select(input,Systematic_Name,Orthogroup) 

mmr.ortho.genes <- mmr.ortho %>%
  filter(!is.na(Orthogroup)) %>%
  dplyr::select(Orthogroup) %>%
  deframe()

mmr.ortho.matrix <- orthogroups %>% 
  filter(Orthogroup %in% mmr.ortho.genes)

```



```{r warning = F}

mmr.phi <- as.data.frame(phi[,which(colnames(phi) %in% mmr.ortho.genes)]) %>% 
  rownames_to_column("Species")
mmr.phi.long <- mmr.phi %>% 
                  dplyr::slice(match(deta.tree.to.use$tip.label,Species)) %>%
                  pivot_longer(-Species,names_to="Orthogroup",values_to="Phi") %>%
               left_join(gc.cont,by="Species")
                 

cor.phi.gc.pic <- mmr.phi.long %>% 
  group_by(Orthogroup) %>% 
  nest(Species,GC,Phi,Clade) %>% 
  mutate(corr = purrr::map(data, function(x)
    {
      tmp <- x %>% 
        dplyr::select(Species,Phi,GC) %>% 
        dplyr::filter(!is.na(Phi) & !is.na(GC)) %>%
        column_to_rownames("Species") %>%
        as.matrix()
      cleaned <- cleanData(deta.tree.to.use,tmp)
      x <- cleaned$data
      tmp.tree <- cleaned$phy
      phi.values <- x[,"Phi"]
      gc.values <- x[,"GC"]
      phi.pic <- calculatePIC(phi.values[tmp.tree$tip.label],tree = tmp.tree)
      gc.pic <- calculatePIC(gc.values[tmp.tree$tip.label],tree=tmp.tree)
      cor.test(phi.pic[,1],gc.pic[,1],method="spearman") 
      
    } %>% tidy())) %>% 
  dplyr::select(-data) %>% 
  unnest(starts_with("corr"),names_sep = "_")

cor.phi.gc.pic <- cor.phi.gc.pic %>% 
  ungroup() %>% 
  group_by(Orthogroup) %>% # how do we correct of multiple hypothesis testing
  mutate(adj_p.value = p.adjust(corr_p.value,method="BH"))


```



```{r fig.width=12,fig.height=6}
correlations.phi.wide <- cor.phi.gc.pic %>% 
                      left_join(mmr.ortho,by="Orthogroup") %>%
                      ungroup() %>%
                      mutate(Significant=ifelse(adj_p.value < 0.05,"*",""))

bar.p.all <- ggplot(correlations.phi.wide,aes(x=input,y=corr_estimate,shape=Significant)) + 
  geom_bar(stat="identity",position = position_dodge(),color="black",fill="white") + 
  geom_point(aes(y = ifelse(corr_estimate > 0, corr_estimate + 0.05, corr_estimate - 0.05)),
             position=position_dodge(0.9),
             show.legend=FALSE) +
  scale_shape_manual(values = c(NA,8)) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ylab("Spearman Correlation (PIC)") +
  xlab("Gene") +
  labs(fill = "Gene Expression-\nGC correlation")
  #ggtitle("Correlation between GC% and MMR gene expression")
bar.p.all

```


```{r}
mmr.phi.mat <- mmr.phi %>%
  dplyr::slice(match(deta.tree.to.use$tip.label, Species)) %>%
  column_to_rownames("Species") %>%
  as.matrix()


```


```{r}
mmr.phi.mat.norm <- mmr.phi.mat %>%
  as.data.frame() %>%
  mutate(across(where(is.numeric),~replace_na(.x,mean(.x,na.rm=T))),
         across(where(is.numeric),~(.x - mean(.x,na.rm=T))/sd(.x,na.rm = T))) %>%
  as.matrix()
 

ppca <- phyl.pca(deta.tree.to.use,mmr.phi.mat.norm)

mmr.scores <-scores(ppca)
mmr.scores <- mmr.scores %>%
  as.data.frame() %>%
  rownames_to_column("Species") %>%
  left_join(gc.cont)




```


```{r fig.width=12,fig.height=12}

gc.p <- ggplot(mmr.scores,aes(x=PC5,y=GC)) +
  geom_point() +
  stat_cor(method="spearman",label.sep = "\n") +
  xlab(paste0("Principal Component 5\n",round(100*ppca$Eval[5,5]),"%")) +
  ylab("GC%") +
  theme_cowplot()
gc.p


load.df <- ppca$L %>%
  as.data.frame() %>%
  rownames_to_column("Orthogroup") %>%
  left_join(mmr.ortho,by="Orthogroup")

load.plot <- ggplot(load.df,aes(x=reorder(input,PC5),y=PC5)) +
  geom_col(fill="white",color="black") +
  theme_cowplot() +
  xlab("Gene") +
  ylab("Principal Component 5 Loadings\n") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  #ggtitle("Principal component loadings for MMR genes")

gc.mmr.phi <- (gc.comp.all | bar.p.all) / (gc.p | load.plot) + plot_annotation(tag_levels = "A")
gc.mmr.phi

ggsave2("../Figures/All_species/gc_content_vs_mmr_expression.pdf",plot=gc.mmr.phi,width=12,height=12)

```